---
title: "E. coli Mice scRNAseq"
subtitle: "Prefiltering QC"
author: "Kennedi Todd"
date: "01/28/2025"
output:
  html_document:
    code_folding: hide
    theme: cerulean
    toc: true
    toc_float: true
editor_options: 
  chunk_output_type: inline
---

# Setup
## Set working directory
```{r working_directory}
knitr::opts_knit$set(root.dir = ".")
```

## Load libraries
```{r libraries, message=FALSE, warnings=FALSE}
library(cowplot)     # plot_grid()
library(dplyr)       # left_join()
library(ggplot2)     # ggplot()
library(gridExtra)   # grid.arrange()
library(parallel)    # detectCores()
library(rtracklayer) # import()
library(scCustomize) # Merge_Seurat_List()
library(Seurat)      # Read10X_h5()
library(stringr)     # str_match()
```

## Variables and functions
```{r set_variables_and_thresholds}
# variables
out <- "../../results/with_integration/pass1/"

# thresholds
nCount.min <- 400
nCount.max <- 20000
nFeature.min <- 250
complexity.cutoff <- 0.8
mt.cutoff <- 10
hb.cutoff <- 1

# single cell functions
files <- list.files("../../functions", full.names = TRUE)
invisible(lapply(files, source))
```

## Load data
```{r read_seurat_obj}
# load data
mouse <- readRDS("../../rObjects/seurat_obj_before_filtering.rds")

# reorder
new_order <- mouse@meta.data %>%
  arrange(treatment, timepoint_days, age, sex, animal_id) %>%
  select(sample)
new_order <- unique(new_order$sample)
mouse$sample <- factor(mouse$sample, levels = new_order)

# reorder columns
new_order <- mouse@meta.data %>%
  arrange(treatment, timepoint_days, age, sex, animal_id) %>%
  rownames()
mouse <- mouse[, new_order]
```

# Pre-filtering QC
## Number of cells
```{r prefiltering_cells_per_sample}
# Visualize the number of cell counts per sample
data <- as.data.frame(table(mouse$sample))
colnames(data) <- c("sample","frequency")

ncells1 <- ggplot(data, aes(x = sample, y = frequency, fill = sample)) + 
  geom_col() +
  theme_bw() +
  geom_text(aes(label = frequency), 
            position=position_dodge(width=0.9), 
            vjust=-0.25) +
  #scale_fill_manual(values = sample_colors) + 
  scale_y_continuous(breaks = seq(0,12000, by = 2000), limits = c(0,12000)) +
  ggtitle("Raw: cells per sample") +
  theme(legend.position =  "none") + 
  theme(axis.text.x = element_text(angle = 45, hjust=1))
ncells1
```

```{r save_prefiltering_ncells}
# save
pdf(paste0(out, "filtering_QC/ncells_raw.pdf"), 
    width = 18, height = 4)
ncells1
dev.off()

# cleanup
remove(ncells1)
```

## Density plots
```{r prefiltering_density, warning=FALSE}
# Visualize nCount_RNA
den1 <- ggplot(mouse@meta.data,
       aes(color = sample,
           x = nCount_RNA,
           fill = sample)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  xlab("nCount_RNA") +
  ylab("Density") +
  theme(legend.position =  "none") +
  geom_vline(xintercept = nCount.min, lty = "dashed") +
  geom_vline(xintercept = nCount.max, lty = "dashed") +
  theme(legend.key.size = unit(0.25, 'cm'), legend.title = element_text(size=9)) +
  scale_x_log10() +
  annotate("text", x = nCount.min, y = 0.1, label = paste("Min:", nCount.min), 
           angle = 0, vjust = -0.5, hjust = -0.1, size = 3) +
  annotate("text", x = nCount.max, y = 0.1, label = paste("Max:", nCount.max), 
           angle = 0, vjust = -0.5, hjust = -0.1, size = 3)

# Visualize nFeature_RNA
den2 <- ggplot(mouse@meta.data,
       aes(color = sample,
           x = nFeature_RNA,
           fill = sample)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_log10() +
  theme(legend.position =  "none") +
  xlab("nFeature_RNA") +
  ylab("Density") +
  geom_vline(xintercept = nFeature.min, lty = "dashed") +
  theme(legend.key.size = unit(0.25, 'cm'), legend.title = element_text(size=9)) +
  annotate("text", x = nFeature.min, y = 0.1, label = paste("Min:", nFeature.min), 
           angle = 0, vjust = -0.5, hjust = -0.1, size = 3)

# Visualize cell complexity
den3 <- ggplot(mouse@meta.data,
       aes(color = sample,
           x = cell_complexity,
           fill = sample)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  #scale_color_manual(values = sample_colors) +
  theme(legend.position =  "none") +
  #scale_fill_manual(values = sample_colors) +
  xlab("Cell Complexity (log10(nFeature/nCount))") +
  ylab("Density") +
  geom_vline(xintercept = complexity.cutoff, lty = "dashed") +
  theme(legend.key.size = unit(0.25, 'cm'), legend.title = element_text(size=9)) +
  annotate("text", x = complexity.cutoff, y = 0.1, label = paste("Min:", complexity.cutoff), 
           angle = 0, vjust = -0.5, hjust = -0.1, size = 3)

# Visualize percent_mt
den4 <- ggplot(mouse@meta.data,
       aes(color = sample,
           x = percent_mt,
           fill = sample)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_continuous(n.breaks = 4) +
  geom_vline(xintercept = mt.cutoff, lty = "dashed") +
  theme(legend.position =  "none") +
  xlab("% Mitochondrial Genes") +
  ylab("Density") +
  theme(legend.key.size = unit(0.25, 'cm'), legend.title = element_text(size=9)) +
  annotate("text", x = mt.cutoff, y = 0.1, label = paste("Min:", mt.cutoff), 
           angle = 0, vjust = -0.5, hjust = -0.1, size = 3)

# Visualize percent_ribo
den5 <- ggplot(mouse@meta.data,
       aes(color = sample,
           x = percent_ribo,
           fill = sample)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_log10() +
  theme(legend.position =  "none") +
  xlab("% Ribosomal Protein Genes") +
  ylab("Density") +
  theme(legend.key.size = unit(0.25, 'cm'), legend.title = element_text(size=9))

# Visualize percent_hb
den6 <- ggplot(mouse@meta.data,
       aes(color = sample,
           x = percent_hb,
           fill = sample)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_log10() +
  geom_vline(xintercept = hb.cutoff, lty = "dashed") +
  theme(legend.position =  "none") +
  xlab("% Hemoglobin Genes") +
  ylab("Density") +
  geom_vline(xintercept = hb.cutoff, lty = "dashed") +
  theme(legend.key.size = unit(0.25, 'cm'), legend.title = element_text(size=9)) +
  annotate("text", x = hb.cutoff, y = 0.1, label = paste("Max:", hb.cutoff), 
           angle = 0, vjust = -0.5, hjust = -0.1, size = 3)

# Arrange graphs in grid
plots <- list(den1,den2,den3,den4,den5,den6)
layout <- rbind(c(1,4),c(2,5),c(3,6))
grid <- grid.arrange(grobs = plots, layout_matrix = layout)
```

```{r save_prefiltering_density,echo=FALSE,eval=FALSE}
# save
pdf(paste0(out, "filtering_QC/density_curves_raw.pdf"), width = 8, height = 8)
grid <- grid.arrange(grobs = plots, layout_matrix = layout)
dev.off()

# cleanup
remove(den1,den2,den3,den4,den5,den6,layout,grid,plots)
```

## Violin plots
```{r prefiltering_violins, warning=FALSE}
# nFeature, nCount, and cell_complexity violins
v1 <- VlnPlot(mouse,
              features = c("nFeature_RNA", "nCount_RNA","cell_complexity"),
              ncol = 1,
              group.by = 'sample',
              raster = FALSE,
              pt.size = 0)
v1

#  percent violins
v2 <- VlnPlot(mouse,
              features = c("percent_mt","percent_ribo","percent_hb"),
              ncol = 1,
              group.by = 'sample',
              raster = FALSE,
              pt.size = 0)
v2
```

```{r save_prefiltering_violins, echo=FALSE, eval=FALSE}
# save v1
pdf(paste0(out, "filtering_QC/violins_nFeature_nCount_complexity_raw.pdf"), 
    width = 12, height = 12)
v1
dev.off()

# save v2
pdf(paste0(out, "filtering_QC/violins_percent_raw.pdf"),
    width = 12, height = 12)
v2
dev.off()

# cleanup
remove(v1,v2)
```

## Scatter plots
```{r prefiltering_scatter, warning=FALSE}
s1 <- ggplot(
  mouse@meta.data,
  aes(x = nCount_RNA, y = nFeature_RNA, color = percent_mt)) + 
  geom_point() + 
  geom_smooth(method = "lm") +
	scale_x_log10() +   	
  scale_y_log10() + 
  theme_classic() +
  geom_vline(xintercept = nCount.min, lty = "dashed", col = "red") + 
  geom_hline(yintercept = nFeature.min, lty = "dashed", col = "red") +
  geom_vline(xintercept = nCount.max, lty = "dashed", col = "red") +
  facet_wrap(~sample) +
  scale_colour_gradient(low = "gray90", high = "black", limits =c(0,100))
s1

s2 <- FeatureScatter(mouse,
                     feature1 = "nCount_RNA",
                     feature2 = "percent_mt",
                     group.by = 'sample',
                     #cols = sample_colors,
                     shuffle = TRUE) +
  geom_hline(yintercept = mt.cutoff, lty = "dashed", col = "red") +
  geom_vline(xintercept = nCount.min, lty = "dashed", col = "red") +
  geom_vline(xintercept = nCount.max, lty = "dashed", col = "red")
s2

s3 <- ggplot(
  mouse@meta.data,
  aes(x = nCount_RNA, y = nFeature_RNA, color = percent_hb)) + 
  geom_point() + 
  geom_smooth(method = "lm") +
	scale_x_log10() +   	
  scale_y_log10() + 
  theme_classic() +
  geom_vline(xintercept = nCount.min, lty = "dashed", col = "red") + 
  geom_hline(yintercept = nFeature.min, lty = "dashed", col = "red") + 
  facet_wrap(~sample) +
  scale_colour_gradient(low = "gray90", high = "black", limits =c(0,100))
s3
```

```{r save_prefiltering_scatter, echo=FALSE, eval=TRUE, warning=FALSE, message=FALSE}
# save
pdf(paste0(out, "filtering_QC/scatter_nFeature_vs_nCount_raw.pdf"),
    width = 18, height = 18)
s1
dev.off()

pdf(paste0(out, "filtering_QC/scatter_percentMT_vs_nCount_raw.pdf"),
    width = 10, height = 6)
s2
dev.off()

# save
pdf(paste0(out, "filtering_QC/scatter_nFeature_vs_nCount_raw2.pdf"),
    width = 18, height = 18)
s3
dev.off()

# cleanup
remove(s1,s2,s3)
```
