---
title: "E. coli Mice scRNAseq"
subtitle: "Clustering and Annotation"
author: "Kennedi Todd"
date: "01/28/2025"
output:
  html_document:
    code_folding: hide
    theme: cerulean
    toc: true
    toc_float: true
editor_options: 
  chunk_output_type: inline
---

# Setup
## Set working directory
```{r working_directory}
knitr::opts_knit$set(root.dir = ".")
```

## Load libraries
```{r libraries, message=FALSE, warnings=FALSE}
# load libraries
library("SeuratObject", lib.loc = "/home/ktodd/R/rstudio-4.3.0-4-with_modules.sif/libs")
library(Azimuth)     # RunAzimuth()
library(cowplot)     # plot_grid()
library(dotenv)
library(dplyr)       # left_join()
library(ggplot2)     # ggplot()
library(gridExtra)   # grid.arrange()
library(parallel)    # detectCores()
library(rtracklayer) # import()
library(scCustomize) # Merge_Seurat_List()
library(Seurat)      # Read10X_h5()
library(stringr)     # str_match()
```

## Variables and functions
```{r set_variables_and_thresholds}
# variables
out <- "../../results/without_integration/"

# single cell functions
files <- list.files("../../functions", full.names = TRUE)
invisible(lapply(files, source))

# load the environment variables
load_dot_env(file = "../../refs/.env")
project_dir <- Sys.getenv("PROJECT_DIR")
```

## Load data
```{r load_data}
mouse.filtered <- readRDS("../../rObjects/seurat_obj_filtered.rds")
```

# Unwanted variation
## Cell cycle
```{r cell_cycle, eval=FALSE}
markers <- paste0(project_dir, "/refs/cell_cycle_markers.tsv")
path <- paste0(project_dir, "/results/without_integration/unwanted_variation")
mouse.filtered[["phase"]] <- cell_cycle_QC(obj = mouse.filtered,
                                           markersPath = markers,
                                           species = "mouse",
                                           outDir = path)
```

## Mitochondria
[Evaluating effects of mitochondrial expression](https://hbctraining.github.io/scRNA-seq_online/lessons/06_SC_SCT_normalization.html)
```{r mitochondria_factor, eval=FALSE}
mouse.filtered[["mito_factor"]] <- 
  mitochondria_QC(obj = mouse.filtered, outDir = path)
```

# Cluster
## SCTransform
- SCTransform method is a more accurate method of normalizing, estimating the 
variance of the raw filtered data, and identifying the most variable genes. \
- Variation in sequencing depth (total nCount_RNA per cell) is normalized using 
a regularized negative binomial model. \
- If there are other sources of uninteresting variation it can be included. \
- NOTE: By default, after normalizing, adjusting the variance, and regressing out 
uninteresting sources of variation, SCTransform will rank the genes by residual 
variance and output the 3000 most variable genes. If the dataset has larger cell 
numbers, then it may be beneficial to adjust this parameter higher using the 
variable.features.n argument. Additionally, the last line of output specifies 
“Set default assay to SCT”. \
- It is suggested to not regress out batch, and instead use a data integration 
method like Harmony \
```{r split_and_scttransform, eval=FALSE}
# split
mouse.filtered[["RNA"]] <- split(mouse.filtered[["RNA"]], 
                                 f = mouse.filtered$sample)
# transform
options(future.globals.maxSize = 8000 * 1024^2)
mouse.filtered <- SCTransform(mouse.filtered, 
                              verbose = FALSE,
                              conserve.memory = TRUE)
```

```{r save_seurat_sct}
saveRDS(mouse.filtered, "../../rObjects/seurat_obj_sct.rds")
```

## PCA
```{r run_pca, warning=FALSE, message=FALSE}
# run PCA on the merged object
mouse.filtered <- RunPCA(object = mouse.filtered, assay = "SCT")
```

```{r plot_pca}
# Reset idents and levels
DefaultAssay(mouse.filtered) <- "SCT"
Idents(mouse.filtered) <- "sample"

# Plot PCA
pca1 <- DimPlot(mouse.filtered,
                reduction = "pca",
                shuffle = TRUE)
pca1
pca2 <- DimPlot(mouse.filtered,
                reduction = "pca",
                split.by = "sample",
                ncol = 8)
pca2

# Plot PC elbow
e1 <- ElbowPlot(mouse.filtered,
                reduction = "pca") +
  geom_vline(xintercept = 13, linetype = "dashed", color = "red")
e1
```

```{r save_pca, echo=FALSE, eval=FALSE, message=FALSE, warning=FALSE}
# save pca1
path <- paste0(out, "clustering_QC/PCA_colored_by_sample.pdf")
pdf(path, width = 12, height = 6)
pca1
dev.off()

# save pca2
path <- paste0(out, "clustering_QC/PCA_split_by_sample.pdf")
pdf(path, width = 16, height = 12)
pca2
dev.off()

# save elbow
path <- paste0(out, "clustering_QC/PCA_elbow.pdf")
pdf(path, width = 6, height = 6)
e1
dev.off()

# cleanup
remove(pca1,pca2,e1)
```

## Integrate
```{r harmony}
# integrate
gc()
mouse.integrated <- IntegrateLayers(object = mouse.filtered,
                                    method = "HarmonyIntegration",
                                    assay = "SCT")

# re-join layers
mouse.integrated[["RNA"]] <- JoinLayers(mouse.integrated[["RNA"]])

# plot harmony
h1 <- DimPlot(mouse.integrated,
              reduction = "harmony",
              shuffle = TRUE,
              raster = FALSE)
h1
h2 <- DimPlot(mouse.integrated,
              reduction = "harmony",
              shuffle = TRUE,
              raster = FALSE,
              ncol = 4,
              split.by = "sample")
h2

# Plot elbow
mouse.integrated@reductions$harmony@stdev <-
  apply(mouse.integrated@reductions$harmony@cell.embeddings, 2, sd)
e1 <- ElbowPlot(mouse.integrated,
                reduction = "harmony") +
  geom_vline(xintercept = 10, linetype = "dashed", color = "red")
e1
```

```{r save_harmony_plots, echo=FALSE, eval=FALSE}
# save h1
path <- paste0(out, "clustering_QC/harmony_colored_by_sample.pdf")
pdf(path, width = 8, height = 6)
h1
dev.off()

# save h2
path <- paste0(out, "clustering_QC/harmony_split_by_sample.pdf")
pdf(path, width = 12, height = 8)
h2
dev.off()

# save e1
path <- paste0(out, "clustering_QC/harmony_elbow.pdf")
pdf(path, width = 6, height = 6)
e1
dev.off()

# cleanup
remove(h1,h2,e1)
```

## UMAP
To overcome the extensive technical noise in the expression of any single gene 
for scRNA-seq data, Seurat assigns cells to clusters based on their PCA scores 
derived from the expression of the integrated most variable genes, with each PC 
essentially representing a “metagene” that combines information across a 
correlated gene set. Determining how many PCs to include in the clustering step 
is therefore important to ensure that we are capturing the majority of the 
variation, or cell types, present in our dataset.
```{r run_UMAP, message=FALSE, warning=FALSE}
# if no integration
mouse.integrated <- mouse.filtered
mouse.integrated[["RNA"]] <- JoinLayers(mouse.integrated[["RNA"]])

# run UMAP
gc()
mouse.integrated <- RunUMAP(mouse.integrated,
                            dims = 1:13,
                            reduction = "pca",
                            n.components = 3) # set to 3 to use with VR

# plot UMAP
DimPlot(mouse.integrated,
        shuffle = TRUE)

# save
saveRDS(mouse.integrated, "../../rObjects/seurat_obj_umap.rds")
remove(mouse.filtered)
```

## Clusters
Seurat uses a graph-based clustering approach, which embeds cells in a graph 
structure, using a K-nearest neighbor (KNN) graph (by default), with edges drawn 
between cells with similar gene expression patterns. Then, it attempts to 
partition this graph into highly interconnected ‘quasi-cliques’ or ‘communities’ 
[Seurat - Guided Clustering Tutorial]. \

We will use the FindClusters() function to perform the graph-based clustering. 
The resolution is an important argument that sets the “granularity” of the 
downstream clustering and will need to be optimized for every individual experiment. 
For datasets of 3,000 - 5,000 cells, the resolution set between 0.4-1.4 generally 
yields good clustering. Increased resolution values lead to a greater number of 
clusters, which is often required for larger datasets. \

The FindClusters() function allows us to enter a series of resolutions and will 
calculate the “granularity” of the clustering. This is very helpful for testing 
which resolution works for moving forward without having to run the function for 
each resolution. \

```{r find_neighbors, message=FALSE, warning=FALSE}
# Determine the K-nearest neighbor graph
mouse.unannotated <- FindNeighbors(object = mouse.integrated,
                                   assay = "SCT",
                                   reduction = "pca",
                                   dims = 1:13)

# Determine the clusters for various resolutions
mouse.unannotated <- FindClusters(object = mouse.unannotated,
                                  algorithm = 1,
                                  resolution = seq(0.1, 0.5, by = 0.1))
mouse.unannotated$seurat_clusters <- mouse.unannotated$SCT_snn_res.0.3
```

## Explore resolutions
```{r explore_resolutions}
# 0.1
DimPlot(mouse.unannotated,
        group.by = "SCT_snn_res.0.1",
        label = TRUE)

# 0.2
DimPlot(mouse.unannotated,
        group.by = "SCT_snn_res.0.2",
        label = TRUE)

# 0.3
DimPlot(mouse.unannotated,
        group.by = "SCT_snn_res.0.3",
        label = TRUE)
```

```{r save_final_object,echo=FALSE,eval=TRUE}
# set params and normalize
Idents(mouse.unannotated) <- "seurat_clusters"
DefaultAssay(mouse.unannotated) <- "RNA"
mouse.unannotated <- NormalizeData(mouse.unannotated)

# save seurat object
saveRDS(mouse.unannotated, "../../rObjects/seurat_obj_unannotated.rds")
remove(mouse.integrated)
```

# Unannotated
## Unannotated UMAP
```{r}
# umap dim 1 & 2
u1 <- DimPlot(mouse.unannotated,
             group.by = "seurat_clusters",
             label = TRUE,
             shuffle = TRUE)
u1

# umap dim 2 & 3
u2 <- DimPlot(mouse.unannotated,
             group.by = "seurat_clusters",
             label = TRUE,
             dims = c(2,3),
             shuffle = TRUE)
u2

# Extract colors
plot_data <- ggplot_build(u1)$data[[1]]
cluster_order <- levels(Idents(mouse.unannotated))
color_mapping <- unique(plot_data[, c("group", "colour")])
color_mapping$group <- factor(color_mapping$group, levels = cluster_order)
color_mapping <- color_mapping[order(color_mapping$group), ]
color_df <- data.frame(Cluster = color_mapping$group, Color = color_mapping$colour)
cluster_colors <- color_df$Color
```

```{r save_unannotated_umap, echo=FALSE, eval=FALSE}
path <- paste0(out, "UMAP/unannotated_clusters_dim1&2.pdf")
pdf(path, width = 8, height = 6)
u1
dev.off()

path <- paste0(out, "UMAP/unannotated_clusters_dim2&3.pdf")
pdf(path, width = 8, height = 6)
u2
dev.off()

remove(u1,u2)
```

## Cells per cluster
```{r cells_per_cluster}
# get num cells per cluster
df <- as.data.frame(table(mouse.unannotated$seurat_clusters))
colnames(df) <- c("cluster","cell_count")

# save
path <- paste0(out, "clustering_QC/cells_per_cluster_unannotated.tsv")
write.table(x = df,
            file = path,
            sep = "\t",
            quote = FALSE,
            row.names = FALSE)
```

## Feature plots
```{r feature_plots, message=FALSE, warning=FALSE}
# UMAP percent_mt
f1 <- FeaturePlot(mouse.unannotated,
            reduction = "umap", 
            features = "percent_mt") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f1

# UMAP percent_ribo
f2 <- FeaturePlot(mouse.unannotated,
            reduction = "umap", 
            features = "percent_ribo") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f2

# UMAP percent_hb
f3 <- FeaturePlot(mouse.unannotated,
            reduction = "umap", 
            features = "percent_hb") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f3

# UMAP nCount
f4 <- FeaturePlot(mouse.unannotated,
            reduction = "umap", 
            features = "nCount_RNA") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f4

# UMAP nFeature
f5 <- FeaturePlot(mouse.unannotated,
            reduction = "umap", 
            features = "nFeature_RNA") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f5

# UMAP cell_complexity
f6 <- FeaturePlot(mouse.unannotated,
            reduction = "umap", 
            features = "cell_complexity") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f6

# UMAP Ttr
f7 <- FeaturePlot(mouse.unannotated,
            reduction = "umap", 
            features = "Ttr") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f7
```

```{r save_feature_plots, echo=FALSE, eval=FALSE}
path <- paste0(out, "clustering_QC/heatmap_")
pdf(paste0(path, "percent_mito.pdf"), width = 8, height = 6)
f1
dev.off()

pdf(paste0(path, "percent_ribo_protein.pdf"), width = 8, height = 6)
f2
dev.off()

pdf(paste0(path, "percent_hb.pdf"), width = 8, height = 6)
f3
dev.off()

pdf(paste0(path, "nCount.pdf"), width = 8, height = 6)
f4
dev.off()

pdf(paste0(path, "nFeature.pdf"), width = 8, height = 6)
f5
dev.off()

pdf(paste0(path, "cell_complexity.pdf"), width = 8, height = 6)
f6
dev.off()

pdf(paste0(path, "choroid.pdf"), width = 8, height = 6)
f7
dev.off()

remove(f1,f2,f3,f4,f5,f6,f7)
```

# Find markers
## Auto find markers
```{r auto_find_markers}
# set params
Idents(mouse.unannotated) <- "seurat_clusters"

# find markers
markers <- SeuratWrappers::RunPrestoAll(
  object = mouse.unannotated,
  assay = "RNA",
  slot = "counts",
  only.pos = FALSE
)

# filter based on adjusted p-value
markers <- markers[markers$p_val_adj < 0.01,]

# get top 3 and top 20 per cluster
top3 <- Reduce(rbind,
               by(markers,
                  markers["cluster"],
                  head,
                  n = 3))
top20 <- Reduce(rbind,
               by(markers,
                  markers["cluster"],
                  head,
                  n = 20))

# save
write.table(markers, 
            paste0(out, "markers/auto_find_markers_adjpval_0.01.tsv"),
            sep = "\t",
            quote = FALSE,
            row.names = FALSE)
```

## Violins
```{r potential_marker_violins}
# auto markers
Idents(mouse.unannotated) <- "seurat_clusters"
v1 <- VlnPlot(mouse.unannotated,
              features = top3$gene,
              split.by = "seurat_clusters",
              group.by = "seurat_clusters",
              cols = cluster_colors,
              stack = TRUE,
              flip = TRUE)
v1

# Microglia, Aif1 aka Iba-1
micro <- c("Aif1","C1qa","C1qb","C1qc","Cx3cr1","Hexb","Itgam","P2ry12","Ptprc",
           "Tmem119")
v2 <- VlnPlot(mouse.unannotated,
             features = micro,
             split.by = "seurat_clusters",
             group.by = "seurat_clusters",
             cols = cluster_colors,
             stack = TRUE,
             flip = TRUE)
v2

astro <- c("Aldoc","Aqp4","Gfap","Gja1","Slc1a2","Slc1a3")
v3 <- VlnPlot(mouse.unannotated,
             features = astro,
             split.by = "seurat_clusters",
             group.by = "seurat_clusters",
             cols = cluster_colors,
             stack = TRUE,
             flip = TRUE)
v3

endo <- c("Cldn5","Flt1","Itm2a","Ly6c1","Pecam1")
v4 <- VlnPlot(mouse.unannotated,
             features = endo,
             split.by = "seurat_clusters",
             group.by = "seurat_clusters",
             cols = cluster_colors,
             stack = TRUE,
             flip = TRUE)
v4

olig <- c("Cldn11","Mag","Mbp","Mog")
v5 <- VlnPlot(mouse.unannotated,
             features = olig,
             split.by = "seurat_clusters",
             group.by = "seurat_clusters",
             cols = cluster_colors,
             stack = TRUE,
             flip = TRUE)
v5

peri <- c("Acta2","Myh11","Rgs5","Pdgfrb")
v6 <- VlnPlot(mouse.unannotated,
             features = peri,
             split.by = "seurat_clusters",
             group.by = "seurat_clusters",
             cols = cluster_colors,
             stack = TRUE,
             flip = TRUE)
v6

fibro <- c("Col1a1","Col1a2","Dcn")
v7 <- VlnPlot(mouse.unannotated,
             features = fibro,
             split.by = "seurat_clusters",
             group.by = "seurat_clusters",
             cols = cluster_colors,
             stack = TRUE,
             flip = TRUE)
v7

poly <- c("Olig2","Tnr","Sox10","Nkx2-2")
v8 <- VlnPlot(mouse.unannotated,
             features = poly,
             split.by = "seurat_clusters",
             group.by = "seurat_clusters",
             cols = cluster_colors,
             stack = TRUE,
             flip = TRUE)
v8

neuro <- c("Eno2","Gad1","Gad2","Syt1","Meg3")
v9 <- VlnPlot(mouse.unannotated,
             features = neuro,
             split.by = "seurat_clusters",
             group.by = "seurat_clusters",
             cols = cluster_colors,
             stack = TRUE,
             flip = TRUE)
v9

v10 <- VlnPlot(mouse.unannotated,
             features =  c("Ttr","Folr1"),
             split.by = "seurat_clusters",
             group.by = "seurat_clusters",
             cols = cluster_colors,
             stack = TRUE,
             flip = TRUE)
v10
```

```{r save_potential_marker_violins}
path <- paste0(out, "markers/unannotated_auto_top3_per_cluster_markers_violin.pdf")
pdf(path, width = 12, height = 14)
v1
dev.off()

path <- paste0(out, "markers/unannotated_microglia_violin.pdf")
pdf(path, width = 10, height = 6)
v2
dev.off()

path <- paste0(out, "markers/unannotated_astrocytes_markers_violin.pdf")
pdf(path, width = 10, height = 4)
v3
dev.off()

path <- paste0(out, "markers/unannotated_endothelial_markers_violin.pdf")
pdf(path, width = 10, height = 4)
v4
dev.off()

path <- paste0(out, "markers/unannotated_oligodendrocyte_markers_violin.pdf")
pdf(path, width = 10, height = 4)
v5
dev.off()

path <- paste0(out, "markers/unannotated_pericytes_smcs_markers_violin.pdf")
pdf(path, width = 10, height = 4)
v6
dev.off()

path <- paste0(out, "markers/unannotated_fibroblast_markers_violin.pdf")
pdf(path, width = 10, height = 4)
v7
dev.off()

path <- paste0(out, "markers/unannotated_polydendrocyte_markers_violin.pdf")
pdf(path, width = 10, height = 4)
v8
dev.off()

path <- paste0(out, "markers/unannotated_neuron_markers_violin.pdf")
pdf(path, width = 10, height = 4)
v9
dev.off()

path <- paste0(out, "markers/unannotated_choroid_markers_violin.pdf")
pdf(path, width = 10, height = 4)
v10
dev.off()

remove(v1,v2,v3,v4,v5,v6,v7,v8,v9)
```

```{r}
# save seurat object
saveRDS(mouse.unannotated, "../../rObjects/seurat_obj_annotated.rds")
```

# Assign identities
```{r assign_identities}
new_cluster_names <- c(
  "0" = "Astrocytes",
  "1" = "Astrocytes",
  "2" = "Microglia",
  "3" = "Microglia",
  "4" = "Endothelial",
  "5" = "Choroid",
  "6" = "Oligodendrocytes",
  "7" = "Endothelial",
  "8" = "Microglia",
  "9" = "Neurons",
  "10" = "Polydendrocytes",
  "11" = "Microglia",
  "12" = "Pericytes & SMCs",
  "13" = "Neurons",
  "14" = "Oligodendrocytes",
  "15" = "Polydendrocytes",
  "16" = "Astrocytes",
  "17" = "Neurogenesis",
  "18" = "Microglia",
  "19" = "Astrocytes",
  "20" = "Astrocytes",
  "21" = "Microglia",
  "22" = "Astrocytes / Endothelial",
  "23" = "?Fibro?Peri?Endo?",
  "24" = "Endothelial",
  "25" = "Endothelial",
  "26" = "Astrocytes",
  "27" = "Microglia",
  "28" = "Choroid"
)

mouse.annotated <- RenameIdents(mouse.unannotated, new_cluster_names)
mouse.annotated$annotated_clusters <- Idents(mouse.annotated)
```

## UMAP
```{r annotated_umap}
cluster_colors <- c("firebrick2", "gold", "chartreuse2", "darkgreen", "cyan",
                    "steelblue", "blue", "violet", "purple3","gray",
                    "chocolate4", "black")
# umap dim 1 & 2
u1 <- DimPlot(mouse.annotated,
             cols = cluster_colors,
             shuffle = TRUE)
u1

# umap dim 2 & 3
u2 <- DimPlot(mouse.annotated,
             dims = c(2,3),
             cols = cluster_colors,
             shuffle = TRUE)
u2
```

```{r save_annotated_umap, echo=FALSE, eval=FALSE}
path <- paste0(out, "UMAP/annotated_clusters_dim1&2.pdf")
pdf(path, width = 8, height = 6)
u1
dev.off()

path <- paste0(out, "UMAP/annotated_clusters_dim2&3.pdf")
pdf(path, width = 8, height = 6)
u2
dev.off()

remove(u1,u2)
```

# Shiny App
- scDRcell is the function used to plot the first UMAP
- inpConf is an argument of scDRcell, inpConf = sc1conf
- infConf gets subset by UI column based on inpdrX which is the user's input for x-axis
- inpdrX is used to set the x-axis label
```{r shiny_app, eval=FALSE}
# create new object
library(ShinyCell)
shiny.obj <- mouse.annotated
VariableFeatures(shiny.obj) <- shiny.obj@assays$SCT@var.features

# set default params
DefaultAssay(shiny.obj) <- "RNA"

# create config
names <- colnames(shiny.obj@meta.data)
names <- names[c(32,33,2:31)]
sc.config <- createConfig(obj = shiny.obj,
                          meta.to.include = names)

# change wd
setwd(out)

# output shiny app folder
makeShinyApp(obj = shiny.obj, 
             scConf = sc.config, 
             gene.mapping = TRUE,
             shiny.title = "E. coli Mice")

# manual config edits
setwd("../../results/without_integration")
sc1conf <- readRDS("shinyApp/sc1conf.rds")
cluster_hex <- "firebrick2|gold|chartreuse2|darkgreen|cyan|steelblue|blue|violet|purple3|gray|chocolate4|black"
sc1conf[2,4] <- cluster_hex
saveRDS(sc1conf, "shinyApp/sc1conf.rds")
```
