---
title: "E. coli Mice scRNAseq"
subtitle: "Postclustering QC"
author: "Kennedi Todd"
date: "01/28/2025"
output:
  html_document:
    code_folding: hide
    theme: cerulean
    toc: true
    toc_float: true
editor_options: 
  chunk_output_type: inline
---

# Setup
## Working directory
```{r working_directory}
knitr::opts_knit$set(root.dir = ".")
```

## Libraries
```{r libraries, message=FALSE, warning=FALSE}
# load packages
library(ComplexUpset) # instersection_size()
library(dplyr)        # ungroup()
library(ggrepel)      # geom_text_repel()
library(gridExtra)    # grid.arrange()
library(gtools)       # smartbind()
library(parallel)     # detectCores()
library(plotly)       # plot_ly()
library(Seurat)       # DimPlot()
library(stringr)      # str_match
library(tidyr)        # %>%
library(UpSetR)       # fromList()
```

## Variables and functions
```{r set_variables_and_functions}
# variables
out <- "../../results/with_integration/pass1/reclusters/group1/"
keep_clusters <- c("Oligodendrocytes","OPCs")

# single cell functions
files <- list.files("../../functions", full.names = TRUE)
invisible(lapply(files, source))
```

## Load data
```{r read_annotated_obj}
mouse.annotated <- readRDS("../../rObjects/seurat_obj_annotated.rds")
```

# Recluster group 1
## Subset
```{r subset_endothelial}
# subset clusters
seurat_subset <- subset(mouse.annotated, annotated_clusters %in% keep_clusters)

# create new seurat object
counts <- GetAssayData(object = seurat_subset, layer = "counts")
seurat_subset <- 
  CreateSeuratObject(counts, meta.data = seurat_subset@meta.data[,c(1:24,31)])

# rename columns
seurat_subset$og_annotations <- seurat_subset$annotated_clusters
seurat_subset$annotated_clusters <- NULL

# check
table(seurat_subset$og_annotations)
```

## Recluster
```{r recluster_endothelial}
# split
seurat_subset[["RNA"]] <- split(seurat_subset[["RNA"]], 
                                 f = seurat_subset$sample)

# transform
options(future.globals.maxSize = 8000 * 1024^2)
seurat_subset <- SCTransform(seurat_subset, 
                              verbose = FALSE)

# PCA
seurat_subset <- RunPCA(object = seurat_subset, assay = "SCT")

# Plot PC elbow
ElbowPlot(seurat_subset, reduction = "pca") +
  geom_vline(xintercept = 10, linetype = "dashed", color = "red")

# join layers
seurat_subset[["RNA"]] <- JoinLayers(seurat_subset[["RNA"]])

# run UMAP
seurat_subset <- RunUMAP(seurat_subset,
                 dims = 1:10,
                 reduction = "pca",
                 n.components = 3) # set to 3 to use with VR

# Determine the K-nearest neighbor graph
seurat_subset <- FindNeighbors(object = seurat_subset,
                                   assay = "SCT",
                                   reduction = "pca",
                                   dims = 1:10)

# Determine the clusters for various resolutions
seurat_subset <- FindClusters(object = seurat_subset.unannotated,
                                  algorithm = 1,
                                  resolution = seq(0.1, 0.5, by = 0.1))
```

## UMAP
```{r}
seurat_subset$seurat_clusters <- seurat_subset$SCT_snn_res.0.2
Idents(seurat_subset) <- "seurat_clusters"
DefaultAssay(seurat_subset) <- "RNA"
seurat_subset <- NormalizeData(seurat_subset)
cluster_colors <- c("firebrick2","coral","cyan","steelblue","blue","green","forestgreen",
                    "gold","violet","purple3","gray","tan","black")

# new clusters
u1 <- DimPlot(seurat_subset,
              cols = cluster_colors,
        raster = FALSE,
        shuffle = TRUE,
        group.by = "seurat_clusters")
u1
u2 <- DimPlot(seurat_subset,
        cols = cluster_colors,
        raster = FALSE,
        shuffle = TRUE,
        dims = c(2,3),
        group.by = "seurat_clusters")
u2

# old annotations
u3 <- DimPlot(seurat_subset,
        raster = FALSE,
        shuffle = TRUE,
        group.by = "og_annotations")
u3
u4 <- DimPlot(seurat_subset,
        raster = FALSE,
        shuffle = TRUE,
        group.by = "og_annotations",
        dims = c(2,3))
u4
```

```{r}
pdf(paste0(out, "new_clusters_umap_dim1&2.pdf"), width = 6, height = 4)
u1
dev.off()

pdf(paste0(out, "new_clusters_umap_dim2&3.pdf"), width = 6, height = 4)
u2
dev.off()

pdf(paste0(out, "old_annotations_umap_dim1&2.pdf"), width = 6, height = 4)
u3
dev.off()

pdf(paste0(out, "old_annotations_umap_dim2&3.pdf"), width = 6, height = 4)
u4
dev.off()

remove(u1,u2,u3,u4)
```

## Heatmap UMAP
```{r feature_plots, warning=FALSE, message=FALSE}
# UMAP percent_mt
f1 <- FeaturePlot(seurat_subset,
                  reduction = "umap", 
                  features = "percent_mt")  + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f1

# UMAP nCount
f2 <- FeaturePlot(seurat_subset,
                  reduction = "umap",
                  features = "nCount_RNA") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f2

# UMAP nFeature
f3 <- FeaturePlot(seurat_subset, 
                  reduction = "umap", 
                  features = "nFeature_RNA") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f3

# UMAP percent_ribo
f4 <- FeaturePlot(seurat_subset,
                  reduction = "umap", 
                  features = "percent_ribo") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f4

# UMAP cell_complexity
f5 <- FeaturePlot(seurat_subset,
                  reduction = "umap",
                  features = "cell_complexity") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f5

# UMAP percent_hb
f6 <- FeaturePlot(seurat_subset, 
                  reduction = "umap",
                  features = "percent_hb") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f6
```

```{r save_feature_plots, echo=FALSE, eval=FALSE}
pdf(paste0(out, "heatmap_percent_mito.pdf"), width = 6, height = 4)
f1
dev.off()

pdf(paste0(out, "heatmap_nCount.pdf"), width = 6, height = 4)
f2
dev.off()

pdf(paste0(out, "heatmap_nFeature.pdf"), width = 6, height = 4)
f3
dev.off()

pdf(paste0(out, "heatmap_percent_ribo.pdf"), width = 6, height = 4)
f4
dev.off()

pdf(paste0(out, "heatmap_cell_complexity.pdf"), width = 6, height = 4)
f5
dev.off()

pdf(paste0(out, "heatmap_percent_hb.pdf"), width = 6, height = 4)
f6
dev.off()

remove(f1,f2,f3,f4,f5,f6)
```

## Auto find markers
```{r auto_find_markers}
# find markers
markers <- SeuratWrappers::RunPrestoAll(
  object = seurat_subset,
  assay = "RNA",
  slot = "counts",
  only.pos = FALSE
)

# filter based on adjusted p-value
markers <- markers[markers$p_val_adj < 0.01,]

# get top 3 per cluster
top3 <- Reduce(rbind,
               by(markers,
                  markers["cluster"],
                  head,
                  n = 3))

v1 <- VlnPlot(seurat_subset,
              features = top3$gene,
              split.by = "seurat_clusters",
              group.by = "seurat_clusters",
              cols = cluster_colors,
              stack = TRUE,
              flip = TRUE)
v1
```

```{r save_auto_markers_violin}
pdf(paste0(out, "auto_markers_violin.pdf"), width = 6, height = 6)
v1
dev.off()
remove(v1,top3,markers)
```

## Manual markers
```{r manual_markers}
v2 <- VlnPlot(seurat_subset,
        features = c("Flt1","Cldn5","Pecam1","Aldoc","Gja1","Aqp4","Pdgfrb",
                     "Myh11","Rgs5","Col1a1","Col1a2"),
        stack = TRUE,
        flip = TRUE,
        cols = cluster_colors,
        split.by = "seurat_clusters")
v2
```

```{r save_manual_markers_violin}
pdf(paste0(out, "manual_markers_violin.pdf"), width = 6, height = 6)
v2
dev.off()
remove(v2)
```


```{r marker_heatmap}
f1 <- FeaturePlot(seurat_subset,
                  reduction = "umap", 
                  features = "Flt1")  + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f1

f2 <- FeaturePlot(seurat_subset,
                  reduction = "umap",
                  features = "Gja1") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f2

f3 <- FeaturePlot(seurat_subset, 
                  reduction = "umap", 
                  features = "Rgs5") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f3

f4 <- FeaturePlot(seurat_subset,
                  reduction = "umap", 
                  features = "Col1a2") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f4
```

```{r save_marker_heatmap}
pdf(paste0(out, "heatmap_Flt1.pdf"), width = 6, height = 4)
f1
dev.off()

pdf(paste0(out, "heatmap_Gja1.pdf"), width = 6, height = 4)
f2
dev.off()

pdf(paste0(out, "heatmap_Rgs5.pdf"), width = 6, height = 4)
f3
dev.off()

pdf(paste0(out, "heatmap_Col1a2.pdf"), width = 6, height = 4)
f4
dev.off()

remove(f1,f2,f3,f4)
```

## Barcodes to remove
```{r barcodes to remove}
barcodes <- seurat_subset@meta.data
barcodes <- barcodes[barcodes$seurat_clusters %in% c(10,12),]
barcodes <- rownames(barcodes)
saveRDS(barcodes, "../../rObjects/group1_remove_barcodes.rds")
remove(seurat_subset)
```

