---
title: "E. coli Mice scRNAseq"
subtitle: "Pass 2 Clustering and Annotation"
author: "Kennedi Todd"
date: "01/28/2025"
output:
  html_document:
    code_folding: hide
    theme: cerulean
    toc: true
    toc_float: true
editor_options: 
  chunk_output_type: inline
---


# Setup
## Working directory
```{r working_directory}
knitr::opts_knit$set(root.dir = ".")
```

## Libraries
```{r libraries, message=FALSE, warning=FALSE}
# load packages
library(ComplexUpset) # instersection_size()
library(dplyr)        # ungroup()
library(ggrepel)      # geom_text_repel()
library(gridExtra)    # grid.arrange()
library(gtools)       # smartbind()
library(parallel)     # detectCores()
library(plotly)       # plot_ly()
library(Seurat)       # DimPlot()
library(stringr)      # str_match
library(tidyr)        # %>%
library(UpSetR)       # fromList()
```

## Variables and functions
```{r set_variables_and_functions}
# variables
out <- "../../results/without_integration/pass2/"

# single cell functions
files <- list.files("../../functions", full.names = TRUE)
invisible(lapply(files, source))
```

## Load data
```{r read_annotated_obj}
mouse.annotated <- readRDS("../../rObjects/seurat_obj_annotated.rds")
```

# Filter
## Cells
Now, from the original mouse.annotated object we filter out the choroid cluster and clusters 10 and 12 from the group 1 recluster.
```{r filter_cells}
group1_barcodes <- readRDS("../../rObjects/group1_remove_barcodes.rds")

mouse.annotated$keep <- !colnames(mouse.annotated) %in% group1_barcodes
mouse <- subset(mouse.annotated, keep == TRUE)
```

## Genes
```{r filter_genes}
mouse <- JoinLayers(mouse)
counts <- GetAssayData(object = mouse, layer = "counts")
nonzero <- counts > 0  # produces logical
keep <- Matrix::rowSums(nonzero) >= 10  # sum the true/false
counts.filtered <- counts[keep,]  # keep certain genes

# create new seurat obj
mouse <- CreateSeuratObject(counts.filtered, 
                            meta.data = mouse@meta.data[c(1,4:24,33,34)])

# rename column
mouse$pass1_annotation <- mouse$annotated_clusters
mouse$annotated_clusters <- NULL

# cleanup
remove(nonzero, mouse.annotated, counts.filtered, counts)
```

# Pass 2
## Recluster
```{r recluster}
# split
mouse[["RNA"]] <- split(mouse[["RNA"]], f = mouse$sample)

# transform
options(future.globals.maxSize = 8000 * 1024^2)
mouse <- SCTransform(mouse, verbose = FALSE)

# PCA
mouse <- RunPCA(object = mouse, assay = "SCT")

# Plot PC elbow
ElbowPlot(mouse, reduction = "pca") +
  geom_vline(xintercept = 10, linetype = "dashed", color = "red")

# join layers
mouse[["RNA"]] <- JoinLayers(mouse[["RNA"]])

# run UMAP
mouse <- RunUMAP(mouse,
                 dims = 1:10,
                 reduction = "pca",
                 n.components = 3) # set to 3 to use with VR

# Determine the K-nearest neighbor graph
mouse.unannotated <- FindNeighbors(object = mouse,
                                   assay = "SCT",
                                   reduction = "pca",
                                   dims = 1:10)

# Determine the clusters for various resolutions
mouse.unannotated <- FindClusters(object = mouse.unannotated,
                                  algorithm = 1,
                                  resolution = seq(0.1, 0.5, by = 0.1))

# set params
mouse.unannotated$seurat_clusters <- mouse.unannotated$SCT_snn_res.0.2
Idents(mouse.unannotated) <- "seurat_clusters"
DefaultAssay(mouse.unannotated) <- "RNA"
mouse.unannotated <- NormalizeData(mouse.unannotated)
saveRDS(mouse.unannotated, "../../rObjects/seurat_obj_unannotated_pass2.rds")
```

## Unnanoted UMAP
```{r unannotated_umap}
# umap dim 1 & 2
u1 <- DimPlot(mouse.unannotated,
              group.by = "seurat_clusters",
              label = TRUE,
              shuffle = TRUE,
              raster = FALSE)
u1

# umap dim 2 & 3
u2 <- DimPlot(mouse.unannotated,
              group.by = "seurat_clusters",
              label = TRUE,
              dims = c(2,3),
              raster = FALSE,
              shuffle = TRUE)
u2

# extract colors
plot_data <- ggplot_build(u1)$data[[1]]
cluster_order <- levels(Idents(mouse.unannotated))
color_mapping <- unique(plot_data[, c("group", "colour")])
color_mapping$group <- factor(color_mapping$group, levels = cluster_order)
color_mapping <- color_mapping[order(color_mapping$group), ]
color_df <- data.frame(Cluster = color_mapping$group, Color = color_mapping$colour)
cluster_colors <- color_df$Color
```

```{r save_unannotated_umap}
path <- paste0(out, "UMAP/unannotated_clusters_dim1&2.pdf")
pdf(path, width = 8, height = 6)
u1
dev.off()

path <- paste0(out, "UMAP/unannotated_clusters_dim2&3.pdf")
pdf(path, width = 8, height = 6)
u2
dev.off()

remove(u1,u2)
```

## Auto find markers
```{r auto_find_markers}
# set params
Idents(mouse.unannotated) <- "seurat_clusters"

# find markers
markers <- SeuratWrappers::RunPrestoAll(
  object = mouse.unannotated,
  assay = "RNA",
  slot = "counts",
  only.pos = FALSE
)

# filter based on adjusted p-value
markers <- markers[markers$p_val_adj < 0.01,]

# get top 3 and top 20 per cluster
top3 <- Reduce(rbind,
               by(markers,
                  markers["cluster"],
                  head,
                  n = 3))

# save
write.table(markers, 
            paste0(out, "markers/auto_find_markers_adjpval_0.01.tsv"),
            sep = "\t",
            quote = FALSE,
            row.names = FALSE)
```

## Violins
```{r potential_marker_violins}
# auto markers
Idents(mouse.unannotated) <- "seurat_clusters"
v1 <- VlnPlot(mouse.unannotated,
              features = top3$gene,
              split.by = "seurat_clusters",
              group.by = "seurat_clusters",
              cols = cluster_colors,
              stack = TRUE,
              flip = TRUE)
v1

# Microglia, Aif1 aka Iba-1
micro <- c("Aif1","C1qa","C1qb","C1qc","Cx3cr1","Hexb","Itgam","P2ry12","Ptprc",
           "Tmem119")
v2 <- VlnPlot(mouse.unannotated,
             features = micro,
             split.by = "seurat_clusters",
             group.by = "seurat_clusters",
             cols = cluster_colors,
             stack = TRUE,
             flip = TRUE)
v2

astro <- c("Aldoc","Aqp4","Gfap","Gja1","Slc1a2","Slc1a3")
v3 <- VlnPlot(mouse.unannotated,
             features = astro,
             split.by = "seurat_clusters",
             group.by = "seurat_clusters",
             cols = cluster_colors,
             stack = TRUE,
             flip = TRUE)
v3

endo <- c("Cldn5","Flt1","Itm2a","Ly6c1","Pecam1")
v4 <- VlnPlot(mouse.unannotated,
             features = endo,
             split.by = "seurat_clusters",
             group.by = "seurat_clusters",
             cols = cluster_colors,
             stack = TRUE,
             flip = TRUE)
v4

olig <- c("Cldn11","Mag","Mbp","Mog")
v5 <- VlnPlot(mouse.unannotated,
             features = olig,
             split.by = "seurat_clusters",
             group.by = "seurat_clusters",
             cols = cluster_colors,
             stack = TRUE,
             flip = TRUE)
v5

peri <- c("Acta2","Myh11","Rgs5","Pdgfrb")
v6 <- VlnPlot(mouse.unannotated,
             features = peri,
             split.by = "seurat_clusters",
             group.by = "seurat_clusters",
             cols = cluster_colors,
             stack = TRUE,
             flip = TRUE)
v6

fibro <- c("Col1a1","Col1a2","Dcn")
v7 <- VlnPlot(mouse.unannotated,
             features = fibro,
             split.by = "seurat_clusters",
             group.by = "seurat_clusters",
             cols = cluster_colors,
             stack = TRUE,
             flip = TRUE)
v7

poly <- c("Olig2","Tnr","Sox10","Nkx2-2")
v8 <- VlnPlot(mouse.unannotated,
             features = poly,
             split.by = "seurat_clusters",
             group.by = "seurat_clusters",
             cols = cluster_colors,
             stack = TRUE,
             flip = TRUE)
v8

neuro <- c("Eno2","Gad1","Gad2","Syt1","Meg3")
v9 <- VlnPlot(mouse.unannotated,
             features = neuro,
             split.by = "seurat_clusters",
             group.by = "seurat_clusters",
             cols = cluster_colors,
             stack = TRUE,
             flip = TRUE)
v9

v10 <- VlnPlot(mouse.unannotated,
             features =  c("Ttr","Folr1"),
             split.by = "seurat_clusters",
             group.by = "seurat_clusters",
             cols = cluster_colors,
             stack = TRUE,
             flip = TRUE)
v10
```

```{r save_potential_marker_violins}
path <- paste0(out, "markers/unannotated_auto_top3_per_cluster_markers_violin.pdf")
pdf(path, width = 12, height = 14)
v1
dev.off()

path <- paste0(out, "markers/unannotated_microglia_violin.pdf")
pdf(path, width = 10, height = 6)
v2
dev.off()

path <- paste0(out, "markers/unannotated_astrocytes_markers_violin.pdf")
pdf(path, width = 10, height = 4)
v3
dev.off()

path <- paste0(out, "markers/unannotated_endothelial_markers_violin.pdf")
pdf(path, width = 10, height = 4)
v4
dev.off()

path <- paste0(out, "markers/unannotated_oligodendrocyte_markers_violin.pdf")
pdf(path, width = 10, height = 4)
v5
dev.off()

path <- paste0(out, "markers/unannotated_pericytes_smcs_markers_violin.pdf")
pdf(path, width = 10, height = 4)
v6
dev.off()

path <- paste0(out, "markers/unannotated_fibroblast_markers_violin.pdf")
pdf(path, width = 10, height = 4)
v7
dev.off()

path <- paste0(out, "markers/unannotated_polydendrocyte_markers_violin.pdf")
pdf(path, width = 10, height = 4)
v8
dev.off()

path <- paste0(out, "markers/unannotated_neuron_markers_violin.pdf")
pdf(path, width = 10, height = 4)
v9
dev.off()

path <- paste0(out, "markers/unannotated_choroid_markers_violin.pdf")
pdf(path, width = 10, height = 4)
v10
dev.off()

remove(v1,v2,v3,v4,v5,v6,v7,v8,v9,v10)
```

# Assign identities
```{r assign_identities}
new_cluster_names <- c(
  "0" = "Astrocytes",
  "1" = "Microglia",
  "2" = "Astrocytes",
  "3" = "Endothelial",
  "4" = "Microglia",
  "5" = "Choroid",
  "6" = "Oligodendrocytes",
  "7" = "Microglia",
  "8" = "Microglia",
  "9" = "Neurons",
  "10" = "Neurons",
  "11" = "OPCs",
  "12" = "Pericytes and SMCs",
  "13" = "Oligodendrocytes",
  "14" = "OPCs",
  "15" = "Microglia",
  "16" = "Microglia",
  "17" = "Astrocytes",
  "18" = "Microglia",
  "19" = "Microglia",
  "20" = "Choroid",
  "21" = "Endothelial",
  "22" = "Astrocytes",
  "23" = "Choroid",
  "24" = "Microglia"
)

mouse.annotated <- RenameIdents(mouse.unannotated, new_cluster_names)
mouse.annotated$annotated_clusters <- Idents(mouse.annotated)
```

## UMAP
```{r annotated_umap}
cluster_colors <- c("firebrick2", "gold", "chartreuse2", "darkgreen", "cyan",
                    "steelblue", "blue", "violet")
# umap dim 1 & 2
u1 <- DimPlot(mouse.annotated,
             cols = cluster_colors,
             raster = FALSE,
             shuffle = TRUE)
u1

# umap dim 2 & 3
u2 <- DimPlot(mouse.annotated,
             dims = c(2,3),
             raster = FALSE,
             cols = cluster_colors,
             shuffle = TRUE)
u2
```

```{r save_annotated_umap, echo=FALSE, eval=FALSE}
path <- paste0(out, "UMAP/annotated_clusters_dim1&2.pdf")
pdf(path, width = 8, height = 6)
u1
dev.off()

path <- paste0(out, "UMAP/annotated_clusters_dim2&3.pdf")
pdf(path, width = 8, height = 6)
u2
dev.off()

remove(u1,u2)
```
## Cluster markers
```{r cluster_workers}
v <- VlnPlot(object = mouse.annotated,
             features = c("Aldoc", "Gja1","C1qa","Tmem119","Cldn5","Flt1",
                          "Ttr","Folr1","Cldn11","Mog","Syt1","Gad1", "Olig2",
                          "Tnr","Rgs5","Pdgfrb"),
             stack = TRUE,
             flip = TRUE,
             cols = cluster_colors,
             group.by = "annotated_clusters",
             split.by = "annotated_clusters")
v
```

```{r save_annotated_cluster_markers, echo=FALSE, eval=FALSE}
pdf(paste0(out, "markers/annotated_cluster_markers.pdf"), width = 8, height = 12)
v
dev.off()
```

```{r save_annotated_obj, echo=FALSE, eval=FALSE}
# save seurat object
saveRDS(mouse.annotated, "../../rObjects/seurat_obj_annotated_pass2.rds")
```

# Shiny App
- scDRcell is the function used to plot the first UMAP
- inpConf is an argument of scDRcell, inpConf = sc1conf
- infConf gets subset by UI column based on inpdrX which is the user's input for x-axis
- inpdrX is used to set the x-axis label
```{r shiny_app, eval=FALSE}
# create new object
library(ShinyCell)
shiny.obj <- mouse.annotated
VariableFeatures(shiny.obj) <- shiny.obj@assays$SCT@var.features

# set default params
DefaultAssay(shiny.obj) <- "RNA"

# create config
names <- colnames(shiny.obj@meta.data)
names <- names[c(34,35,2:33)]
sc.config <- createConfig(obj = shiny.obj,
                          meta.to.include = names)

# change wd
setwd(out)

# output shiny app folder
makeShinyApp(obj = shiny.obj, 
             scConf = sc.config, 
             gene.mapping = TRUE,
             shiny.title = "E. coli Mice")

# manual config edits
setwd(out)
sc1conf <- readRDS("shinyApp/sc1conf.rds")
cluster_hex <- "firebrick2|gold|chartreuse2|darkgreen|cyan|steelblue|blue|violet"
sc1conf[2,4] <- cluster_hex
saveRDS(sc1conf, "shinyApp/sc1conf.rds")
```
