---
title: "E. coli Mice scRNAseq"
subtitle: "Pass 3: SoupX and Recluster"
author: "Kennedi Todd"
date: "12/01/2025"
output:
  html_document:
    code_folding: hide
    theme: cerulean
    toc: true
    toc_float: true
editor_options: 
  chunk_output_type: inline
---

# Setup
## Set working directory
```{r working_directory}
knitr::opts_knit$set(root.dir = ".")
```

## Load libraries
```{r libraries, message=FALSE, warnings=FALSE}
library(dotenv)         # load_dot_env()
library(dplyr)          # %>%
library(ggplot2)        # geom_vline()
library(gridExtra)      # grid.arrange()
library(Seurat)         # DimPlot()
library(SeuratWrappers) # RunPrestoAll()
library(SoupX)          # SoupChannel()
library(tidyr)          # pivot_wider()
```

## Variables and functions
```{r set_variables_and_thresholds}
# source thresholds and output paths
source("../../refs/thresholds_and_outs.R")
out <- paste0(out, "pass3/")

# single cell functions
files <- list.files("../../functions", full.names = TRUE)
invisible(lapply(files, source))

# load the environment variables
load_dot_env(file = "../../refs/.env")
project_dir <- Sys.getenv("PROJECT_DIR")
```

## Load data
```{r}
mouse_annotated <- readRDS("../../rObjects/pass2_annotated_seurat_obj.rds")
```

## UMAP
```{r}
cluster_colors <- get_cluster_colors(
  num_clusters = length(unique(mouse_annotated$annotated_clusters)))

DimPlot(mouse_annotated, 
        shuffle = TRUE,
        cols = cluster_colors)
```

# SoupX 
## rhoPrior 0.2
```{r soupx_auto_default}
# store the corrected counts for each sample
corrected_list <- list()
seq_ids <- unique(mouse_annotated$tgen_subject_name)
sample_ids <- unique(mouse_annotated$sample_id)
rho_estimates <- setNames(numeric(length(sample_ids)), sample_ids)

for (i in seq_along(seq_ids)) {

  cat("\nProcessing sample:", sample_ids[i], "\n")
  
  # get table of droplets (tod) from Cell Ranger output
  # genes as rows, cells as columns
  cellranger_out_dir <- 
    paste0(project_dir, "/counts/", seq_ids[i], "/outs/multi/count/raw_feature_bc_matrix")
  tod <- Seurat::Read10X(cellranger_out_dir)
  colnames(tod) <- paste0(sample_ids[i], "_", colnames(tod))
  
  # get table of counts (toc) from filtered Seurat object
  # cells are a subset of tod
  toc <- mouse_annotated@assays$RNA$counts[, mouse_annotated$sample_id == sample_ids[i]]
  
  # subset tod genes to match genes in toc
  genes_to_keep <- rownames(toc)
  tod <- tod[genes_to_keep, ]
  stopifnot(all(rownames(tod) == rownames(toc)))
  
  # build SoupChannel manually
  cat("\n\t Building SoupChannel \n")
  sc <- SoupChannel(tod, toc)

  # get cluster labels for toc
  all.equal(colnames(toc), rownames(sc$metaData))
  toc_clusters <- mouse_annotated$annotated_clusters[colnames(toc)]
  
  # add to SoupChannel
  sc <- setClusters(sc, toc_clusters)
  sc <- setDR(sc, mouse_annotated@reductions$umap@cell.embeddings[colnames(toc), ])
  
  # estimate
  cat("\t Estimating rho \n")
  sc <- autoEstCont(sc, priorRho = 0.2)
  
  # save top contaminators
  cat("\t Saving top 200 contaminators \n")
  top_contam <- sc$soupProfile[order(sc$soupProfile$est, decreasing = TRUE),]
  top_contam$est <- round(top_contam$est, digits = 4)
  top_contam <- top_contam[1:200,]
  top_contam$gene <- rownames(top_contam)
  write.table(x = top_contam,
              file = paste0(out, "soupx/", sample_ids[i], "_top200_contaminators.tsv"),
              sep = "\t",
              quote = FALSE,
              row.names = FALSE)
  
  # adjust (auto mode with rounding)
  cat("\t Adjusting counts \n")
  sc.cleaned <- adjustCounts(sc, roundToInt = TRUE)

  # save corrected counts
  corrected_list[[sample_ids[i]]] <- sc.cleaned
  cat("Done:", sample_ids[i], "\n")
  
  # save rho estimate
  rho_estimates[sample_ids[i]] <- sc$fit$rhoEst
}

cat("\nAll samples processed & merged!\n")

# save the corrected list
saveRDS(corrected_list,
        paste0("../../rObjects/pass3_soupx.rds"),
        compress = FALSE)

# merge corrected counts
corrected_counts <- do.call(cbind, corrected_list)

# save rho estimates
write.table(
  x = data.frame(Sample = names(rho_estimates), Rho = rho_estimates),
  file = paste0(out, "soupx/rho_estimates.tsv"),
  sep = "\t",
  quote = FALSE,
  row.names = FALSE
)
```

## Non-expressed list
```{r}
# define marker genes
Astrocyte <- c("Aldoc","Aqp4")
Choroid <- c("Ttr","Folr1")
Endothelial <- c("Flt1", "Cldn5", "Ptprb")
Fibroblasts <- c("Col1a1","Col1a2","Dcn")
Microglia <- c("Tmem119","P2ry12","C1qa")
BAM <- c("Mrc1","C1qa","Apoe")
PanNeurons <- c("Snap25", "Syt1", "Meg3")
ExcitatoryNeurons <- c("Slc16a6", "Slc16a7")
InhibitoryNeurons <- c("Gad1", "Gad2")
AllNeurons <- c(PanNeurons, ExcitatoryNeurons, InhibitoryNeurons)
Oligodendrocyte <- c("Mbp","Mog","Mag","Opalin")
OPCs <- c("Pdgfra","Cspg4")

# non-expressed gene list for every cluster
nonExpressedGeneList <- list(
  `Astrocytes` = unique(c(Choroid, Endothelial, Fibroblasts, Microglia,
                          AllNeurons, Oligodendrocyte, OPCs)),
  `Choroid` = unique(c(Astrocyte, Endothelial, Fibroblasts, Microglia, BAM, 
                       AllNeurons, Oligodendrocyte, OPCs)),
  `Endothelial` = unique(c(Astrocyte, Choroid, Fibroblasts, Microglia, BAM,
                           AllNeurons, Oligodendrocyte, OPCs)),
  `Fibroblasts` = unique(c(Astrocyte, Choroid, Endothelial, Microglia, BAM,
                           AllNeurons, Oligodendrocyte, OPCs)),
  `Microglia` = unique(c(Astrocyte, Choroid, Endothelial, Fibroblasts, AllNeurons,
                         Oligodendrocyte, OPCs)),
  `BAM` = unique(c(Astrocyte, Choroid, Endothelial, Fibroblasts, AllNeurons,
                   Oligodendrocyte, OPCs)),
  `Excitatory Neurons` = unique(c(Astrocyte, Choroid, Endothelial, Fibroblasts, 
                                  Microglia, BAM, Oligodendrocyte, OPCs, 
                                  InhibitoryNeurons)),
  `Inhibitory Neurons` = unique(c(Astrocyte, Choroid, Endothelial, Fibroblasts,
                                  Microglia, BAM, Oligodendrocyte, OPCs, 
                                  ExcitatoryNeurons)),
  `Oligodendrocytes` = unique(c(Astrocyte, Choroid, Endothelial, Fibroblasts, 
                                Microglia, BAM, AllNeurons, OPCs)),
  `OPCs` = unique(c(Astrocyte, Choroid, Endothelial, Fibroblasts, Microglia, 
                    BAM, AllNeurons, Oligodendrocyte))
)
```

## Loop
```{r}
# store the corrected counts for each sample
corrected_list <- list()
seq_ids <- unique(mouse_annotated$root_filename)
sample_ids <- unique(mouse_annotated$sample)
rho_estimates <- setNames(numeric(length(sample_ids)), sample_ids)

for (i in seq_along(seq_ids)) {

  cat("\nProcessing sample", i, "\n")
  
  # get table of droplets (tod) from Cell Ranger output
  # genes as rows, droplets (empty + cells) as columns
  cellranger_out_dir <- paste0(project_dir, "/counts/", seq_ids[i], 
                               "/outs/raw_feature_bc_matrix")
  tod <- Seurat::Read10X(cellranger_out_dir)
  colnames(tod) <- paste0(sample_ids[i], "_", colnames(tod))
  
  # get table of counts (toc) from filtered Seurat object
  # cells are a subset of tod
  toc <- mouse_annotated@assays$RNA$counts[, mouse_annotated$sample == sample_ids[i], 
                                           drop = FALSE]
  
  # align genes
  genes_to_keep <- intersect(rownames(tod), rownames(toc))
  tod <- tod[genes_to_keep, , drop = FALSE]
  toc <- toc[genes_to_keep, , drop = FALSE]
  
  # build SoupChannel manually
  cat("\n\t Building SoupChannel \n")
  sc <- SoupChannel(tod, toc, calcSoupProfile = FALSE)
  
  # get cluster labels
  all.equal(colnames(toc), rownames(sc$metaData))
  toc_clusters <- mouse_annotated$annotated_clusters[colnames(toc)]
  sc <- setClusters(sc, toc_clusters)
  
  # add dimensional reduction
  sc <- setDR(sc, mouse_annotated@reductions$umap@cell.embeddings[colnames(toc), ])
  
  # Estimate contamination using your manual gene lists
  sc <- estimateSoup(sc = sc, 
                     soupRange = c(0, 100), 
                     keepDroplets = FALSE)
  useToEst <- estimateNonExpressingCells(sc,
                                         nonExpressedGeneList = nonExpressedGeneList,
                                         clusters = toc_clusters)
  sc <- calculateContaminationFraction(sc,
                                       nonExpressedGeneList = nonExpressedGeneList,
                                       useToEst = useToEst)

  # save top contaminators
  cat("\t Saving top 200 contaminators \n")
  top_contam <- sc$soupProfile[order(sc$soupProfile$est, decreasing = TRUE),]
  top_contam$est <- round(top_contam$est, digits = 4)
  top_contam <- top_contam[1:200,]
  top_contam$gene <- rownames(top_contam)
  write.table(x = top_contam,
              file = paste0(out, "soupx/", sample_ids[i], "_top200_contaminators.tsv"),
              sep = "\t",
              quote = FALSE,
              row.names = FALSE)
  
  # adjust counts
  cat("\t Adjusting counts \n")
  sc.cleaned <- adjustCounts(sc, roundToInt = TRUE)

  # save corrected counts
  corrected_list[[sample_ids[i]]] <- sc.cleaned
  cat("Done:", sample_ids[i], "\n")
  
  # save rho estimate
  rho_estimates[sample_ids[i]] <- unique(sc$metaData$rho)
}

cat("\nAll samples processed & merged!\n")

# save the corrected list
saveRDS(corrected_list,
        "../../rObjects/pass3_soupx.rds",
        compress = FALSE)

# merge corrected counts
corrected_counts <- do.call(cbind, corrected_list)

# save rho estimates
write.table(
  x = data.frame(Sample = names(rho_estimates), Rho = rho_estimates),
  file = paste0(out, "soupx/rho_estimates.tsv"),
  sep = "\t",
  quote = FALSE,
  row.names = FALSE
)

# cleanup
remove(corrected_list, nonExpressedGeneList, sc, sc.cleaned, toc, tod,
       top_contam, useToEst)
```

# Recluster pass 3
## Create seurat obj
```{r}
# filter lowly expressed genes
nonzero <- corrected_counts > 0  # produces logical
keep <- Matrix::rowSums(nonzero) >= 10  # sum the true/false
counts_filtered <- corrected_counts[keep,]  # keep certain genes

# create new Seurat object with corrected counts
options(Seurat.object.assay.calcn = TRUE)
mouse <- CreateSeuratObject(counts = counts_filtered)

# add metadata
mouse <- AddMetaData(mouse, 
                     metadata = mouse_annotated@meta.data[colnames(mouse), 4:24])

# print features removed
print(paste0(dim(corrected_counts)[1] - dim(counts_filtered)[1], " features removed"))

# cleanup data
remove(corrected_counts, counts_filtered, nonzero, mouse_annotated)
```

## SCTransform
```{r sctransform}
# split
mouse_split <- SplitObject(mouse, split.by = "sample")

# transform
options(future.globals.maxSize = 20000 * 1024^2)  # 20 GB
mouse_split <- lapply(mouse_split, SCTransform)

# get shared featured
shared_features <- SelectIntegrationFeatures(mouse_split)

# merge
mouse_merged <- merge(mouse_split[[1]], y = mouse_split[-1])

# set default assay
DefaultAssay(mouse_merged) <- "SCT"

# set variable features
VariableFeatures(mouse_merged) <- shared_features

# save
saveRDS(mouse_merged, 
        file = "../../rObjects/pass3_sct_seurat_obj.rds",
        compress = FALSE)
remove(mouse, mouse_split)
gc()
```

## PCA
```{r run_pca, warning=FALSE, message=FALSE}
# run PCA
mouse_merged <- RunPCA(object = mouse_merged, assay = "SCT")
```

```{r plot_pca}
pca1 <- DimPlot(mouse_merged,
                reduction = "pca",
                group.by = "treatment",
                shuffle = TRUE)
pca1

pca2 <- DimPlot(mouse_merged,
                reduction = "pca",
                group.by = "sample",
                shuffle = TRUE)
pca2

pca3 <- DimPlot(mouse_merged,
                reduction = "pca",
                group.by = "group",
                shuffle = TRUE)
pca3

pca4 <- DimPlot(mouse_merged,
                reduction = "pca",
                group.by = "group2",
                shuffle = TRUE)
pca4

pca5 <- DimPlot(mouse_merged,
                reduction = "pca",
                group.by = "age",
                shuffle = TRUE)
pca5

pca6 <- DimPlot(mouse_merged,
                reduction = "pca",
                group.by = "timepoint_days",
                shuffle = TRUE)
pca6

pca7 <- DimPlot(mouse_merged,
                reduction = "pca",
                group.by = "sex",
                shuffle = TRUE)
pca7

# PC elbow
e1 <- ElbowPlot(mouse_merged,
                ndims = 40,
                reduction = "pca") +
  geom_vline(xintercept = 13, linetype = "dashed", color = "red")
e1
```

```{r save_pca, message=FALSE, warning=FALSE}
path <- paste0(out, "clustering_QC/PCA_colored_by_treatment.pdf")
pdf(path, width = 6, height = 4)
pca1
dev.off()

path <- paste0(out, "clustering_QC/PCA_colored_by_sample.pdf")
pdf(path, width = 8, height = 4)
pca2
dev.off()

path <- paste0(out, "clustering_QC/PCA_colored_by_group.pdf")
pdf(path, width = 6, height = 4)
pca3
dev.off()

path <- paste0(out, "clustering_QC/PCA_colored_by_group2.pdf")
pdf(path, width = 6, height = 4)
pca4
dev.off()

path <- paste0(out, "clustering_QC/PCA_colored_by_age.pdf")
pdf(path, width = 6, height = 4)
pca5
dev.off()

path <- paste0(out, "clustering_QC/PCA_colored_by_timepoint.pdf")
pdf(path, width = 6, height = 4)
pca6
dev.off()

path <- paste0(out, "clustering_QC/PCA_colored_by_sex.pdf")
pdf(path, width = 6, height = 4)
pca7
dev.off()

# save elbow
path <- paste0(out, "clustering_QC/PCA_elbow.pdf")
pdf(path, width = 6, height = 6)
e1
dev.off()

# cleanup
remove(pca1,pca2,pca3,pca4,pca5,pca6,pca7,e1)
```

## Harmony
```{r harmony}
# Integrate
Idents(mouse_merged) <- "sample"
mouse_integrated <- IntegrateLayers(object = mouse_merged,
                                    method = "HarmonyIntegration",
                                    assay = "SCT")

# save
saveRDS(mouse_integrated, 
        file = "../../rObjects/pass3_harmony_seurat_obj.rds",
        compress = FALSE)
remove(mouse_merged)
gc()
```

```{r plot_harmony}
# Plot harmony
harmony1 <- DimPlot(mouse_integrated,
                reduction = "harmony",
                group.by = "treatment",
                shuffle = TRUE)
harmony1

harmony2 <- DimPlot(mouse_integrated,
                reduction = "harmony",
                group.by = "sample",
                shuffle = TRUE)
harmony2

harmony3 <- DimPlot(mouse_integrated,
                    reduction = "harmony",
                    group.by = "group",
                    shuffle = TRUE)
harmony3

harmony4 <- DimPlot(mouse_integrated,
                    reduction = "harmony",
                    group.by = "group2",
                    shuffle = TRUE)
harmony4

harmony5 <- DimPlot(mouse_integrated,
                    reduction = "harmony",
                    group.by = "age",
                    shuffle = TRUE)
harmony5

harmony6 <- DimPlot(mouse_integrated,
                    reduction = "harmony",
                    group.by = "timepoint_days",
                    shuffle = TRUE)
harmony6

harmony7 <- DimPlot(mouse_integrated,
                    reduction = "harmony",
                    group.by = "sex",
                    shuffle = TRUE)
harmony7
```

```{r save_harmony, message=FALSE, warning=FALSE}
path <- paste0(out, "clustering_QC/harmony_colored_by_treatment.pdf")
pdf(path, width = 6, height = 4)
harmony1
dev.off()

path <- paste0(out, "clustering_QC/harmony_colored_by_sample.pdf")
pdf(path, width = 8, height = 4)
harmony2
dev.off()

path <- paste0(out, "clustering_QC/harmony_colored_by_group.pdf")
pdf(path, width = 6, height = 4)
harmony3
dev.off()

path <- paste0(out, "clustering_QC/harmony_colored_by_group2.pdf")
pdf(path, width = 6, height = 4)
harmony4
dev.off()

path <- paste0(out, "clustering_QC/harmony_colored_by_age.pdf")
pdf(path, width = 6, height = 4)
harmony5
dev.off()

path <- paste0(out, "clustering_QC/harmony_colored_by_timepoint.pdf")
pdf(path, width = 6, height = 4)
harmony6
dev.off()

path <- paste0(out, "clustering_QC/harmony_colored_by_sex.pdf")
pdf(path, width = 6, height = 4)
harmony7
dev.off()

# cleanup
remove(harmony1,harmony2,harmony3,harmony4,harmony5,harmony6,harmony7)
```

## UMAP and Cluster
```{r umap_and_cluster, message=FALSE, warning=FALSE}
# run UMAP
mouse_unannotated <- RunUMAP(mouse_integrated,
                             dims = 1:27,
                             min.dist = 0.2,
                             n.neighbors = 70,
                             reduction = "harmony",
                             n.components = 3)

# Determine the K-nearest neighbor graph
mouse_unannotated <- FindNeighbors(object = mouse_unannotated,
                                   assay = "SCT",
                                   reduction = "harmony",
                                   dims = 1:27)

# Determine the clusters for various resolutions
mouse_unannotated <- FindClusters(object = mouse_unannotated,
                                  algorithm = 1,
                                  resolution = 0.3)

# Set default assay and normalize for visualization
mouse_unannotated <- JoinLayers(mouse_unannotated, assay = "RNA")
DefaultAssay(mouse_unannotated) <- "RNA"
mouse_unannotated <- NormalizeData(mouse_unannotated)

# save
saveRDS(mouse_unannotated, 
        file = "../../rObjects/pass3_unannotated_seurat_obj.rds",
        compress = FALSE)
```

# Unannotated pass 3
## UMAP
```{r unannotated_umap}
Idents(mouse_unannotated) <- "seurat_clusters"

u1 <- DimPlot(mouse_unannotated, shuffle = TRUE, label = TRUE)
u1
```

```{r save_unannotated_umap}
pdf(paste0(out, "UMAP/unannotated_umap.pdf"), width = 6, height = 5)
u1
dev.off()
remove(u1)
```


## Canonical markers
```{r unannotated_canonical_markers}
# Canonical
canonical_markers <- c(
  "Aif1", "C1qa", "P2ry12", "Tmem119","Trem2","Mrc1", # Microglia, BAM
  "Cd3e","Cd4","Cd8a","Itgax","S100a9",               # Immune (T, B, DC, neutrophil)
  "Aldoc", "Aqp4", "Gja1", "Clu", "Apoe", "Slc1a3",   # Astrocytes
  "Ptprb", "Cldn5", "Flt1",                           # Endothelial
  "Pdgfrb", "Rgs5", "Acta2", "Vtn",                   # Mural
  "Syt1", "Snap25", "Meg3", "Kcnip4",                 # Neurons
  "Gad1", "Gad2", "Slc17a6", "Slc17a7",               # Neurons
  "Col1a1", "Col1a2", "Dcn",                          # Fibroblast
  "Pdgfra", "Cspg4", "Olig1", "Olig2", "Tnr",         # OPCs
  "Mog", "Opalin", "Mal", "Mag", "Mbp",               # Oligodendrocytes
   "Ttr", "Folr1", "Cfap44", "Spag16"                 # Choroid
)

d2 <- DotPlot(mouse_unannotated,
              features = canonical_markers,
              cols = "RdYlBu",
              assay = "RNA") +
  coord_flip() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) 
d2
```

```{r save_unannotated_canonical_markers}
pdf(paste0(out, "markers/unannotated_canonical_markers.pdf"), width = 10, height = 10)
d2
dev.off()
remove(d2)
```

## Auto markers
```{r unannotated_auto_markers}
auto_markers <- SeuratWrappers::RunPrestoAll(
  object = mouse_unannotated,
  assay = "RNA",
  slot = "counts",
  only.pos = TRUE
)

top2 <- auto_markers %>%
  filter(p_val_adj < 0.01) %>%
  group_by(cluster) %>%
  slice_head(n = 2) %>%
  ungroup()

dot <- DotPlot(mouse_unannotated,
               features = unique(top2$gene),
               cols = "RdYlBu",
               assay = "RNA") +
  coord_flip() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) 
dot
```

```{r save_unannotated_auto_markers}
pdf(paste0(out, "markers/unannotated_auto_markers.pdf"), width = 10, height = 10)
dot
dev.off()
remove(dot, top2)
```

## Assign idetities
```{r assign_identities}
new_cluster_names <- c(
  "0" = "Microglia",
  "1" = "Endothelial",
  "2" = "Astrocytes",
  "3" = "Astrocytes",
  "4" = "Oligodendrocytes",
  "5" = "Microglia",
  "6" = "Astrocytes",
  "7" = "OPCs",
  "8" = "Choroid",
  "9" = "Neurons 1",
  "10" = "Neurons 2",
  "11" = "Oligodendrocytes",
  "12" = "Astrocytes",
  "13" = "Microglia",
  "14" = "BAM",
  "15" = "Fibroblasts",
  "16" = "Neurons 1"
)

# set new idents
mouse_annotated <- RenameIdents(mouse_unannotated, new_cluster_names)
mouse_annotated$annotated_clusters <- Idents(mouse_annotated)

# set level
mouse_annotated$annotated_clusters <- factor(mouse_annotated$annotated_clusters,
                                  levels = c("Astrocytes",
                                             "Choroid",
                                             "Endothelial",
                                             "Fibroblasts",
                                             "Microglia",
                                             "BAM",
                                             "Neurons 1",
                                             "Neurons 2",
                                             "OPCs",
                                             "Oligodendrocytes"))
Idents(mouse_annotated) <- "annotated_clusters"

# save
saveRDS(mouse_annotated, 
        file = "../../rObjects/pass3_annotated_seurat_obj.rds",
        compress = FALSE)
```

# Annotated pass 3
## UMAP
```{r annotated_umap}
cluster_colors <- c(
  "Astrocytes"  = "#F0E442",
  "Choroid"     = "#FDB863",
  "Endothelial" = "#E08214",
  "Fibroblasts" = "#B35806",
  "Microglia" = "#542788",
  "BAM" = "#998EC3",
  "Neurons 1" = "#0868AC",
  "Neurons 2" = "#2B8CBE",
  "Oligodendrocytes" = "#009E73",
  "OPCs" = "#7FCDBB"
)

u1 <- DimPlot(mouse_annotated, 
              group.by = "annotated_clusters", 
              cols = cluster_colors,
              shuffle = TRUE)
u1
```

```{r save_annotated_umap}
pdf(paste0(out, "UMAP/annotated_umap.pdf"), width = 6, height = 4)
u1
dev.off()
remove(u1,u2)
```

## Canonical markers
```{r annotated_canonical_markers}
# Canonical
canonical_markers <- c(
  "Aldoc", "Aqp4", "Gja1", "Clu", "Apoe",             # Astrocytes
  "Cfap44", "Spag16", "Ttr", "Folr1",                 # Choroid
  "Ptprb", "Cldn5", "Flt1",                           # Endothelial
  "Pdgfrb", "Rgs5", "Acta2", "Vtn",                   # Mural
  "Col1a1", "Col1a2", "Dcn",                          # Fibroblast
  "Aif1", "C1qa", "P2ry12", "Tmem119","Trem2","Mrc1", # Microglia
  "Syt1", "Snap25", "Rbfox3", "Slc17a6", "Slc17a7",   # Neurons
  "Stmn2", "Tubb3", "Map2", "Gad1","Gad2",            # Neurons
  "Pdgfra", "Cspg4", "Olig1", "Olig2", "Tnr",         # OPCs
  "Mog", "Opalin", "Mal", "Mag", "Mbp"                # Oligodendrocytes
)

d2 <- DotPlot(mouse_annotated,
              features = canonical_markers,
              cols = "RdYlBu",
              assay = "RNA") +
  coord_flip() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) 
d2
```

```{r save_annotated_markers}
pdf(paste0(out, "markers/annotated_canonical_markers.pdf"), width = 8, height = 10)
d2
dev.off()
remove(d2)
```


# QC
## Heatmap UMAP
```{r umap_heatmap, warning=FALSE, message=FALSE}
# UMAP percent_mt
f1 <- FeaturePlot(mouse_annotated, 
            reduction = "umap", 
            features = "percent_mt")  + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f1

# UMAP nCount
f2 <- FeaturePlot(mouse_annotated, 
            reduction = "umap",
            features = "nCount_RNA") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f2

# UMAP nFeature
f3 <- FeaturePlot(mouse_annotated, 
            reduction = "umap", 
            features = "nFeature_RNA") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f3

# UMAP percent_ribo
f4 <- FeaturePlot(mouse_annotated, 
            reduction = "umap", 
            features = "percent_ribo") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f4

# UMAP cell_complexity
f5 <- FeaturePlot(mouse_annotated, 
            reduction = "umap", 
            features = "cell_complexity") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f5

# UMAP percent_hb
f6 <- FeaturePlot(mouse_annotated, 
            reduction = "umap", 
            features = "percent_hb") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f6
```

```{r save_umap_heatmap, echo=FALSE, eval=FALSE}
path <- paste0(out, "clustering_QC/umap_heatmap_")
pdf(paste0(path, "percent_mito.pdf"), width = 8, height = 6)
f1
dev.off()

pdf(paste0(path, "nCount.pdf"), width = 8, height = 6)
f2
dev.off()

pdf(paste0(path, "nFeature.pdf"), width = 8, height = 6)
f3
dev.off()

pdf(paste0(path, "percent_ribo.pdf"), width = 8, height = 6)
f4
dev.off()

pdf(paste0(path, "cell_complexity.pdf"), width = 8, height = 6)
f5
dev.off()

pdf(paste0(path, "percent_hb.pdf"), width = 8, height = 6)
f6
dev.off()

remove(f1,f2,f3,f4,f5,f6)
```

## Percent cells per cluster
```{r percent_cells}
# timepoint
b1 <- mouse_annotated@meta.data %>%
  group_by(annotated_clusters, timepoint_days) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=timepoint_days)) +
  theme_classic() +
  geom_col() +
  ggtitle("Percentage of time point per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b1

# sex
b2 <- mouse_annotated@meta.data %>%
  group_by(annotated_clusters, sex) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=sex)) +
  theme_classic() +
  geom_col() +
  ggtitle("Percentage of sex per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b2

# sample
b3 <- mouse_annotated@meta.data %>%
  group_by(annotated_clusters, sample) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=sample)) +
  theme_classic() +
  geom_col() +
  ggtitle("Percentage of sample per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b3

# phase
b4 <- mouse_annotated@meta.data %>%
  group_by(annotated_clusters, phase) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=phase)) +
  theme_classic() +
  geom_col() +
  ggtitle("Percentage of phase per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b4

# mito.factor
b5 <- mouse_annotated@meta.data %>%
  group_by(annotated_clusters, mito_factor) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=mito_factor)) +
  theme_classic() +
  geom_col() +
  ggtitle("Percentage of mito factor per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b5

# group
b6 <- mouse_annotated@meta.data %>%
  group_by(annotated_clusters, group) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=group)) +
  theme_classic() +
  geom_col() +
  ggtitle("Percentage of group per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b6

# group2
b7 <- mouse_annotated@meta.data %>%
  group_by(annotated_clusters, group2) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=group2)) +
  theme_classic() +
  geom_col() +
  ggtitle("Percentage of group2 per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b7

# age
b8 <- mouse_annotated@meta.data %>%
  group_by(annotated_clusters, age) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=age)) +
  theme_classic() +
  geom_col() +
  ggtitle("Percentage of age per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b8

# treatment
b9 <- mouse_annotated@meta.data %>%
  group_by(annotated_clusters, treatment) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=treatment)) +
  theme_classic() +
  geom_col() +
  ggtitle("Percentage of treatment per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b9
```

```{r save_percent_cells,echo=FALSE,eval=FALSE}
# save
path <- paste0(out, "clustering_QC/percent_")
pdf(paste0(path, "timepoint_per_cluster.pdf"), width = 8, height = 6)
b1
dev.off()

# save
pdf(paste0(path, "sex_per_cluster.pdf"), width = 8, height = 6)
b2
dev.off()

# save
pdf(paste0(path, "sample_per_cluster.pdf"), width = 8, height = 8)
b3
dev.off()

# save
pdf(paste0(path, "phase_per_cluster.pdf"), width = 8, height = 6)
b4
dev.off()

# save
pdf(paste0(path, "mito_factor_per_cluster.pdf"), width = 8, height = 6)
b5
dev.off()

# save
pdf(paste0(path, "group_per_cluster.pdf"), width = 8, height = 6)
b6
dev.off()

# save
pdf(paste0(path, "group2_per_cluster.pdf"), width = 8, height = 6)
b7
dev.off()

# save
pdf(paste0(path, "age_per_cluster.pdf"), width = 8, height = 6)
b8
dev.off()

# save
pdf(paste0(path, "treatment_per_cluster.pdf"), width = 8, height = 6)
b9
dev.off()

# cleanup
remove(b1,b2,b3,b4,b5,b6,b7,b8,b9)
```

## Cells per sample
```{r cells_per_sample}
data <- as.data.frame(table(mouse_annotated$sample))
colnames(data) <- c("sample","frequency")

ncells <- ggplot(data, aes(x = sample, y = frequency, fill = sample)) + 
  geom_col() +
  theme_classic() +
  geom_text(aes(label = frequency), 
            position=position_dodge(width=0.9), 
            vjust=-0.25) +
  scale_y_continuous(breaks = seq(0,8000, by = 2000), limits = c(0,8000)) +
  ggtitle("Filtered: cells per sample") +
  theme(legend.position =  "none") + 
  theme(axis.text.x = element_text(angle = 45, hjust=1))
ncells
```

```{r save_cells_per_sample, echo=FALSE, eval=FALSE}
pdf(paste0(out, "clustering_QC/cells_per_sample.pdf"), width = 18, height = 4)
ncells
dev.off()
```

## Cells per cluster
```{r cells_per_cluster}
data <- as.data.frame(table(mouse_annotated$annotated_clusters))
colnames(data) <- c("cluster","frequency")

ncells <- ggplot(data, aes(x = cluster, y = frequency, fill = cluster)) + 
  geom_col() +
  theme_classic() +
  geom_text(aes(label = frequency), 
            position=position_dodge(width=0.9), 
            vjust=-0.25) +
  scale_y_continuous(breaks = seq(0,40000, by = 10000), limits = c(0,40000)) +
  ggtitle("Cells per Cluster") +
  scale_fill_manual(values = cluster_colors) +
  theme(legend.position =  "none") + 
  theme(axis.text.x = element_text(angle = 45, hjust=1))
ncells
```

```{r save_cells_per_cluster, echo=FALSE, eval=FALSE}
pdf(paste0(out, "clustering_QC/cells_per_cluster.pdf"), width = 8, height = 6)
ncells
dev.off()
```

## Cells per cluster per sample
```{r cells_per_cluster_per_sample}
# extract data
data <- data.frame(cluster = mouse_annotated$annotated_clusters,
                   sample = mouse_annotated$sample)
data_summary <- data %>%
  group_by(cluster, sample) %>%
  summarise(count = n(), .groups = 'drop') %>%
  ungroup()

# Create the stacked bar plot
ncells <- ggplot(data_summary, aes(x = cluster, y = count, fill = sample)) +
  geom_bar(stat = "identity") +
  labs(y = "Frequency", x = "Cluster", fill = "Sample") +
  theme_minimal() +
  ggtitle("Frequency of Cells in Each Cluster with Sample Distribution") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust=1)) +
  ylim(c(0,40000))
ncells

# reformat for table output
data_summary <- data_summary %>% pivot_wider(names_from = cluster, values_from = count)
```

```{r save_cells_per_cluster_per_sample, echo=FALSE, eval=FALSE}
pdf(paste0(out, "clustering_QC/cells_per_cluster_with_sample_distribution.pdf"),
    width = 12, height = 6)
ncells
dev.off()
```

## Cells per cluster per group
```{r cells_per_cluster_per_group}
# extract data
data <- data.frame(cluster = mouse_annotated$annotated_clusters,
                   group = mouse_annotated$group)
data_summary <- data %>%
  group_by(cluster, group) %>%
  summarise(count = n(), .groups = 'drop') %>%
  ungroup()

# Create the stacked bar plot
ncells <- ggplot(data_summary, aes(x = cluster, y = count, fill = group)) +
  geom_bar(stat = "identity") +
  labs(y = "Frequency", x = "Cluster", fill = "Group") +
  theme_minimal() +
  ggtitle("Frequency of Cells in Each Cluster with Group Distribution") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust=1)) +
  ylim(c(0,40000))
ncells

# reformat for table output
data_summary <- data_summary %>% pivot_wider(names_from = cluster, values_from = count)
```

```{r save_cells_per_cluster_per_group, echo=FALSE, eval=FALSE}
pdf(paste0(out, "clustering_QC/cells_per_cluster_with_group_distribution.pdf"),
    width = 8, height = 6)
ncells
dev.off()

write.table(data_summary, 
            file = paste0(out, "clustering_QC/cells_per_cluster_with_group_distribution.tsv"), 
            sep = "\t",quote = FALSE, row.names = FALSE)
```


