---
title: "E. coli Mice scRNAseq"
subtitle: "Recluster Microglia"
author: "Kennedi Todd"
date: "12/30/2025"
output:
  html_document:
    code_folding: hide
    theme: cerulean
    toc: true
    toc_float: true
editor_options: 
  chunk_output_type: inline
---

# Setup
## Set working directory
```{r working_directory}
knitr::opts_knit$set(root.dir = ".")
```

## Load libraries
```{r libraries, message=FALSE, warnings=FALSE}
library(dplyr)          # %>%
library(ggplot2)        # geom_vline()
library(gridExtra)      # grid.arrange()
library(Seurat)         # DimPlot()
library(SeuratWrappers) # RunPrestoAll()
library(tidyr)          # pivot_wider()
```

## Variables and functions
```{r set_variables_and_thresholds}
# source thresholds and output paths
source("../../refs/thresholds_and_outs.R")
out <- paste0(out, "pass3/microglia_recluster/")

# single cell functions
files <- list.files("../../functions", full.names = TRUE)
invisible(lapply(files, source))
```

## Load data
```{r}
mouse_annotated <- readRDS("../../rObjects/pass3_annotated_seurat_obj.rds")
```

## UMAP
```{r}
cluster_colors <- c(
  "Astrocytes"  = "#F0E442",
  "Choroid"     = "#FDB863",
  "Endothelial" = "#E08214",
  "Fibroblasts" = "#B35806",
  "Microglia" = "#542788",
  "BAM" = "#998EC3",
  "Neurons 1" = "#0868AC",
  "Neurons 2" = "#2B8CBE",
  "Oligodendrocytes" = "#009E73",
  "OPCs" = "#7FCDBB"
)

DimPlot(mouse_annotated,
        cols = cluster_colors,
        shuffle = TRUE)
```

# Recluster microglia
## Subset
```{r subset_seurat}
# subset on cell types of interest
mouse <- subset(mouse_annotated, annotated_clusters == "Microglia")

# check
table(mouse$annotated_clusters)
```

## Filter genes
```{r filter_genes}
# filter genes
counts <- GetAssayData(object = mouse, layer = "counts")
nonzero <- counts > 0  # produces logical
keep <- Matrix::rowSums(nonzero) >= 10  # sum the true/false
counts_filtered <- counts[keep,]  # keep certain genes

# overwrite sub
all.equal(colnames(counts_filtered), rownames(mouse@meta.data))
mouse <- CreateSeuratObject(counts_filtered)
mouse <- AddMetaData(mouse, metadata = mouse_annotated@meta.data[colnames(mouse), 4:24])

# print features removed
print(paste0(dim(counts)[1] - dim(counts_filtered)[1], " features removed"))

# cleanup data
remove(counts,counts_filtered,nonzero,mouse_annotated)
```

## SCTransform
```{r sctransform}
# split
mouse_split <- SplitObject(mouse, split.by = "sample")

# transform
options(future.globals.maxSize = 20000 * 1024^2)  # 20 GB
mouse_split <- lapply(mouse_split, SCTransform)

# get shared featured
shared_features <- SelectIntegrationFeatures(mouse_split)

# merge
mouse_merged <- merge(mouse_split[[1]], y = mouse_split[-1])

# set default assay
DefaultAssay(mouse_merged) <- "SCT"

# set variable features
VariableFeatures(mouse_merged) <- shared_features

# save
saveRDS(mouse_merged, 
        file = "../../rObjects/microglia_recluster_sct_seurat_obj.rds",
        compress = FALSE)
remove(mouse, mouse_split)
gc()
```

## PCA
```{r run_pca, warning=FALSE, message=FALSE}
# run PCA
mouse_merged <- RunPCA(object = mouse_merged, assay = "SCT")
```

```{r plot_pca}
pca1 <- DimPlot(mouse_merged,
                reduction = "pca",
                group.by = "treatment",
                shuffle = TRUE)
pca1

pca2 <- DimPlot(mouse_merged,
                reduction = "pca",
                group.by = "sample",
                shuffle = TRUE)
pca2

pca3 <- DimPlot(mouse_merged,
                reduction = "pca",
                group.by = "group",
                shuffle = TRUE)
pca3

pca4 <- DimPlot(mouse_merged,
                reduction = "pca",
                group.by = "group2",
                shuffle = TRUE)
pca4

pca5 <- DimPlot(mouse_merged,
                reduction = "pca",
                group.by = "age",
                shuffle = TRUE)
pca5

pca6 <- DimPlot(mouse_merged,
                reduction = "pca",
                group.by = "timepoint_days",
                shuffle = TRUE)
pca6

pca7 <- DimPlot(mouse_merged,
                reduction = "pca",
                group.by = "sex",
                shuffle = TRUE)
pca7

# PC elbow
e1 <- ElbowPlot(mouse_merged,
                ndims = 40,
                reduction = "pca") +
  geom_vline(xintercept = 10, linetype = "dashed", color = "red")
e1
```

```{r save_pca, message=FALSE, warning=FALSE}
path <- paste0(out, "clustering_QC/PCA_colored_by_treatment.pdf")
pdf(path, width = 6, height = 4)
pca1
dev.off()

path <- paste0(out, "clustering_QC/PCA_colored_by_sample.pdf")
pdf(path, width = 8, height = 4)
pca2
dev.off()

path <- paste0(out, "clustering_QC/PCA_colored_by_group.pdf")
pdf(path, width = 6, height = 4)
pca3
dev.off()

path <- paste0(out, "clustering_QC/PCA_colored_by_group2.pdf")
pdf(path, width = 6, height = 4)
pca4
dev.off()

path <- paste0(out, "clustering_QC/PCA_colored_by_age.pdf")
pdf(path, width = 6, height = 4)
pca5
dev.off()

path <- paste0(out, "clustering_QC/PCA_colored_by_timepoint.pdf")
pdf(path, width = 6, height = 4)
pca6
dev.off()

path <- paste0(out, "clustering_QC/PCA_colored_by_sex.pdf")
pdf(path, width = 6, height = 4)
pca7
dev.off()

# save elbow
path <- paste0(out, "clustering_QC/PCA_elbow.pdf")
pdf(path, width = 6, height = 6)
e1
dev.off()

# cleanup
remove(pca1,pca2,pca3,pca4,pca5,pca6,pca7,e1)
```

## UMAP and Cluster
```{r umap_and_cluster, message=FALSE, warning=FALSE}
# run UMAP
mouse_unannotated <- RunUMAP(mouse_merged,
                             dims = 1:10,
                             reduction = "pca",
                             n.components = 3)

# Determine the K-nearest neighbor graph
mouse_unannotated <- FindNeighbors(object = mouse_unannotated,
                                   assay = "SCT",
                                   reduction = "pca",
                                   dims = 1:10)

# Determine the clusters for various resolutions
mouse_unannotated <- FindClusters(object = mouse_unannotated,
                                  algorithm = 1,
                                  resolution = c(0.1, 0.2, 0.3))

# Set default assay and normalize for visualization
mouse_unannotated <- JoinLayers(mouse_unannotated, assay = "RNA")
DefaultAssay(mouse_unannotated) <- "RNA"
mouse_unannotated <- NormalizeData(mouse_unannotated)

# prepend
mouse_unannotated$seurat_clusters <- paste0("cluster", mouse_unannotated$seurat_clusters)

# save
saveRDS(mouse_unannotated, 
        file = "../../rObjects/microglia_recluster_unannotated_seurat_obj.rds",
        compress = FALSE)
```

# Unannotated
## UMAP
```{r unannotated_umap}
Idents(mouse_unannotated) <- "seurat_clusters"

u1 <- DimPlot(mouse_unannotated, shuffle = TRUE, label = TRUE)
u1
```

```{r save_unannotated_umap}
pdf(paste0(out, "UMAP/unannotated_umap.pdf"), width = 6, height = 5)
u1
dev.off()
remove(u1)
```

## Canonical markers
```{r unannotated_canonical_markers}
# Canonical
canonical_markers <- c(
  "Aif1", "C1qa", "P2ry12", "Tmem119","Trem2","Mrc1", # Microglia, BAM
  "Cd3e","Cd4","Cd8a","Itgax","S100a9",               # Immune (T, B, DC, neutrophil)
  "Aldoc", "Aqp4", "Gja1", "Clu", "Apoe", "Slc1a3",   # Astrocytes
  "Ptprb", "Cldn5", "Flt1",                           # Endothelial
  "Pdgfrb", "Rgs5", "Acta2", "Vtn",                   # Mural
  "Syt1", "Snap25", "Meg3", "Kcnip4",                 # Neurons
  "Gad1", "Gad2", "Slc17a6", "Slc17a7",               # Neurons
  "Col1a1", "Col1a2", "Dcn",                          # Fibroblast
  "Pdgfra", "Cspg4", "Olig1", "Olig2", "Tnr",         # OPCs
  "Mog", "Opalin", "Mal", "Mag", "Mbp",               # Oligodendrocytes
   "Ttr", "Folr1", "Cfap44", "Spag16"                 # Choroid
)

d2 <- DotPlot(mouse_unannotated,
              features = canonical_markers,
              cols = "RdYlBu",
              assay = "RNA") +
  coord_flip() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) 
d2
```

```{r save_unannotated_canonical_markers}
pdf(paste0(out, "markers/unannotated_canonical_markers.pdf"), width = 10, height = 10)
d2
dev.off()
remove(d2)
```

## Auto markers
```{r unannotated_auto_markers}
auto_markers <- SeuratWrappers::RunPrestoAll(
  object = mouse_unannotated,
  assay = "RNA",
  slot = "counts",
  only.pos = TRUE
)

top5 <- auto_markers %>%
  filter(p_val_adj < 0.01) %>%
  group_by(cluster) %>%
  slice_head(n = 5) %>%
  ungroup()

dot <- DotPlot(mouse_unannotated,
               features = unique(top5$gene),
               cols = "RdYlBu",
               assay = "RNA") +
  coord_flip() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) 
dot
```

```{r save_unannotated_auto_markers}
write.table(auto_markers,
            file = paste0(out, "markers/unannotated_auto_markers.tsv"),
            sep = "\t",
            quote = FALSE,
            col.names = TRUE,
            row.names = FALSE)

pdf(paste0(out, "markers/unannotated_auto_top5_markers.pdf"), width = 10, height = 10)
dot
dev.off()
remove(dot, top5)
```


# QC
## Heatmap UMAP
```{r umap_heatmap, warning=FALSE, message=FALSE}
# UMAP percent_mt
f1 <- FeaturePlot(mouse_unannotated, 
            reduction = "umap", 
            features = "percent_mt")  + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f1

# UMAP nCount
f2 <- FeaturePlot(mouse_unannotated, 
            reduction = "umap",
            features = "nCount_RNA") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f2

# UMAP nFeature
f3 <- FeaturePlot(mouse_unannotated, 
            reduction = "umap", 
            features = "nFeature_RNA") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f3

# UMAP percent_ribo
f4 <- FeaturePlot(mouse_unannotated, 
            reduction = "umap", 
            features = "percent_ribo") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f4

# UMAP cell_complexity
f5 <- FeaturePlot(mouse_unannotated, 
            reduction = "umap", 
            features = "cell_complexity") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f5

# UMAP percent_hb
f6 <- FeaturePlot(mouse_unannotated, 
            reduction = "umap", 
            features = "percent_hb") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f6
```

```{r save_umap_heatmap, echo=FALSE, eval=FALSE}
path <- paste0(out, "clustering_QC/umap_heatmap_")
pdf(paste0(path, "percent_mito.pdf"), width = 8, height = 6)
f1
dev.off()

pdf(paste0(path, "nCount.pdf"), width = 8, height = 6)
f2
dev.off()

pdf(paste0(path, "nFeature.pdf"), width = 8, height = 6)
f3
dev.off()

pdf(paste0(path, "percent_ribo.pdf"), width = 8, height = 6)
f4
dev.off()

pdf(paste0(path, "cell_complexity.pdf"), width = 8, height = 6)
f5
dev.off()

pdf(paste0(path, "percent_hb.pdf"), width = 8, height = 6)
f6
dev.off()

remove(f1,f2,f3,f4,f5,f6)
```

## Percent cells per cluster
```{r percent_cells}
# timepoint
b1 <- mouse_unannotated@meta.data %>%
  group_by(seurat_clusters, timepoint_days) %>%
  dplyr::count() %>%
  group_by(seurat_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=seurat_clusters,y=percent, fill=timepoint_days)) +
  theme_classic() +
  geom_col() +
  ggtitle("Percentage of time point per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b1

# sex
b2 <- mouse_unannotated@meta.data %>%
  group_by(seurat_clusters, sex) %>%
  dplyr::count() %>%
  group_by(seurat_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=seurat_clusters,y=percent, fill=sex)) +
  theme_classic() +
  geom_col() +
  ggtitle("Percentage of sex per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b2

# sample
b3 <- mouse_unannotated@meta.data %>%
  group_by(seurat_clusters, sample) %>%
  dplyr::count() %>%
  group_by(seurat_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=seurat_clusters,y=percent, fill=sample)) +
  theme_classic() +
  geom_col() +
  ggtitle("Percentage of sample per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b3

# phase
b4 <- mouse_unannotated@meta.data %>%
  group_by(seurat_clusters, phase) %>%
  dplyr::count() %>%
  group_by(seurat_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=seurat_clusters,y=percent, fill=phase)) +
  theme_classic() +
  geom_col() +
  ggtitle("Percentage of phase per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b4

# mito.factor
b5 <- mouse_unannotated@meta.data %>%
  group_by(seurat_clusters, mito_factor) %>%
  dplyr::count() %>%
  group_by(seurat_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=seurat_clusters,y=percent, fill=mito_factor)) +
  theme_classic() +
  geom_col() +
  ggtitle("Percentage of mito factor per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b5

# group
b6 <- mouse_unannotated@meta.data %>%
  group_by(seurat_clusters, group) %>%
  dplyr::count() %>%
  group_by(seurat_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=seurat_clusters,y=percent, fill=group)) +
  theme_classic() +
  geom_col() +
  ggtitle("Percentage of group per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b6

# group2
b7 <- mouse_unannotated@meta.data %>%
  group_by(seurat_clusters, group2) %>%
  dplyr::count() %>%
  group_by(seurat_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=seurat_clusters,y=percent, fill=group2)) +
  theme_classic() +
  geom_col() +
  ggtitle("Percentage of group2 per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b7

# age
b8 <- mouse_unannotated@meta.data %>%
  group_by(seurat_clusters, age) %>%
  dplyr::count() %>%
  group_by(seurat_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=seurat_clusters,y=percent, fill=age)) +
  theme_classic() +
  geom_col() +
  ggtitle("Percentage of age per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b8

# treatment
b9 <- mouse_unannotated@meta.data %>%
  group_by(seurat_clusters, treatment) %>%
  dplyr::count() %>%
  group_by(seurat_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=seurat_clusters,y=percent, fill=treatment)) +
  theme_classic() +
  geom_col() +
  ggtitle("Percentage of treatment per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b9
```

```{r save_percent_cells,echo=FALSE,eval=FALSE}
# save
path <- paste0(out, "clustering_QC/percent_")
pdf(paste0(path, "timepoint_per_cluster.pdf"), width = 8, height = 6)
b1
dev.off()

# save
pdf(paste0(path, "sex_per_cluster.pdf"), width = 8, height = 6)
b2
dev.off()

# save
pdf(paste0(path, "sample_per_cluster.pdf"), width = 8, height = 8)
b3
dev.off()

# save
pdf(paste0(path, "phase_per_cluster.pdf"), width = 8, height = 6)
b4
dev.off()

# save
pdf(paste0(path, "mito_factor_per_cluster.pdf"), width = 8, height = 6)
b5
dev.off()

# save
pdf(paste0(path, "group_per_cluster.pdf"), width = 8, height = 6)
b6
dev.off()

# save
pdf(paste0(path, "group2_per_cluster.pdf"), width = 8, height = 6)
b7
dev.off()

# save
pdf(paste0(path, "age_per_cluster.pdf"), width = 8, height = 6)
b8
dev.off()

# save
pdf(paste0(path, "treatment_per_cluster.pdf"), width = 8, height = 6)
b9
dev.off()

# cleanup
remove(b1,b2,b3,b4,b5,b6,b7,b8,b9)
```

## Cells per sample
```{r cells_per_sample}
data <- as.data.frame(table(mouse_unannotated$sample))
colnames(data) <- c("sample","frequency")

ncells <- ggplot(data, aes(x = sample, y = frequency, fill = sample)) + 
  geom_col() +
  theme_classic() +
  geom_text(aes(label = frequency), 
            position=position_dodge(width=0.9), 
            vjust=-0.25) +
  scale_y_continuous(breaks = seq(0,2000, by = 500), limits = c(0,2000)) +
  ggtitle("Filtered: cells per sample") +
  theme(legend.position =  "none") + 
  theme(axis.text.x = element_text(angle = 45, hjust=1))
ncells
```

```{r save_cells_per_sample, echo=FALSE, eval=FALSE}
pdf(paste0(out, "clustering_QC/cells_per_sample.pdf"), width = 18, height = 4)
ncells
dev.off()
```

## Cells per cluster
```{r cells_per_cluster}
data <- as.data.frame(table(mouse_unannotated$seurat_clusters))
colnames(data) <- c("cluster","frequency")

ncells <- ggplot(data, aes(x = cluster, y = frequency, fill = cluster)) + 
  geom_col() +
  theme_classic() +
  geom_text(aes(label = frequency), 
            position=position_dodge(width=0.9), 
            vjust=-0.25) +
  scale_y_continuous(breaks = seq(0,12000, by = 3000), limits = c(0,12000)) +
  ggtitle("Cells per Cluster") +
  scale_fill_manual(values = cluster_colors) +
  theme(legend.position =  "none") + 
  theme(axis.text.x = element_text(angle = 45, hjust=1))
ncells
```

```{r save_cells_per_cluster, echo=FALSE, eval=FALSE}
pdf(paste0(out, "clustering_QC/cells_per_cluster.pdf"), width = 8, height = 6)
ncells
dev.off()
```

## Cells per cluster per sample
```{r cells_per_cluster_per_sample}
# extract data
data <- data.frame(cluster = mouse_unannotated$seurat_clusters,
                   sample = mouse_unannotated$sample)
data_summary <- data %>%
  group_by(cluster, sample) %>%
  summarise(count = n(), .groups = 'drop') %>%
  ungroup()

# Create the stacked bar plot
ncells <- ggplot(data_summary, aes(x = cluster, y = count, fill = sample)) +
  geom_bar(stat = "identity") +
  labs(y = "Frequency", x = "Cluster", fill = "Sample") +
  theme_minimal() +
  ggtitle("Frequency of Cells in Each Cluster with Sample Distribution") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust=1)) +
  ylim(c(0,12000))
ncells

# reformat for table output
data_summary <- data_summary %>% pivot_wider(names_from = cluster, values_from = count)
```

```{r save_cells_per_cluster_per_sample, echo=FALSE, eval=FALSE}
pdf(paste0(out, "clustering_QC/cells_per_cluster_with_sample_distribution.pdf"),
    width = 12, height = 6)
ncells
dev.off()
```

## Cells per cluster per group
```{r cells_per_cluster_per_group}
# extract data
data <- data.frame(cluster = mouse_unannotated$seurat_clusters,
                   group = mouse_unannotated$group)
data_summary <- data %>%
  group_by(cluster, group) %>%
  summarise(count = n(), .groups = 'drop') %>%
  ungroup()

# Create the stacked bar plot
ncells <- ggplot(data_summary, aes(x = cluster, y = count, fill = group)) +
  geom_bar(stat = "identity") +
  labs(y = "Frequency", x = "Cluster", fill = "Group") +
  theme_minimal() +
  ggtitle("Frequency of Cells in Each Cluster with Group Distribution") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust=1)) +
  ylim(c(0,12000))
ncells

# reformat for table output
data_summary <- data_summary %>% pivot_wider(names_from = cluster, values_from = count)
```

```{r save_cells_per_cluster_per_group, echo=FALSE, eval=FALSE}
pdf(paste0(out, "clustering_QC/cells_per_cluster_with_group_distribution.pdf"),
    width = 8, height = 6)
ncells
dev.off()

write.table(data_summary, 
            file = paste0(out, "clustering_QC/cells_per_cluster_with_group_distribution.tsv"), 
            sep = "\t",quote = FALSE, row.names = FALSE)
```


# Pseudobulk DESeq2
- DESeq2 expects raw (integer) counts as input \
- It automatically estimates size factors to normalize for sequencing depth and library size differences between samples. \
- If you pre-normalize (e.g., CPM, TPM, log-transformed), the internal DESeq2 model becomes invalid — you’ll violate its statistical assumptions. \
## Aggregate counts
```{r aggregate_counts}
# sum counts by cell type and sample
pb_counts <- AggregateExpression(
  mouse_unannotated,
  group.by = c("seurat_clusters", "sample"),
  slot = "counts",
  normalization.method = "none",
  return.seurat = FALSE
)$RNA

# get sample meta
sample_meta <- unique(mouse_unannotated@meta.data[, c("sample", "group", "group2", 
                                                    "treatment", "timepoint_days",
                                                    "age", "sex")])
rownames(sample_meta) <- sample_meta$sample
```

## Split matrix
```{r split_matrix}
cell_types <- unique(mouse_unannotated$seurat_clusters)

split_pb_counts <- lapply(cell_types, function(ct) {
  cols <- grep(paste0("^", ct, "_"), colnames(pb_counts), value = TRUE)
  pb_counts[, cols, drop = FALSE]
})

names(split_pb_counts) <- cell_types

remove(pb_counts)
```

## Define comparisons
```{r define_comparisons}
comparisons <- list(
  c("group2", "E.2D.Y.F", "S.2D.Y.F"),
  c("group2", "E.2D.Y.M", "S.2D.Y.M"),
  c("group2", "E.2D.O.F", "S.2D.O.F"),
  c("group2", "E.2D.O.M", "S.2D.O.M"),
  c("group2", "E.21D.Y.F", "S.21D.Y.F"),
  c("group2", "E.21D.Y.M", "S.21D.Y.M"),
  c("group2", "E.21D.O.F", "S.21D.O.F"),
  c("group2", "E.21D.O.M", "S.21D.O.M"),
  
  c("group2", "E.21D.Y.F", "E.2D.Y.F"),
  c("group2", "E.21D.Y.M", "E.2D.Y.M"),
  c("group2", "E.21D.O.F", "E.2D.O.F"),
  c("group2", "E.21D.O.M", "E.2D.O.M"),
  c("group2", "S.21D.Y.F", "S.2D.Y.F"),
  c("group2", "S.21D.Y.M", "S.2D.Y.M"),
  c("group2", "S.21D.O.F", "S.2D.O.F"),
  c("group2", "S.21D.O.M", "S.2D.O.M"),
  
  c("group2", "E.2D.O.F", "E.2D.Y.F"),
  c("group2", "E.2D.O.M", "E.2D.Y.M"),
  c("group2", "S.2D.O.F", "S.2D.Y.F"),
  c("group2", "S.2D.O.M", "S.2D.Y.M"),
  c("group2", "E.21D.O.F", "E.21D.Y.F"),
  c("group2", "E.21D.O.M", "E.21D.Y.M"),
  c("group2", "S.21D.O.F", "S.21D.Y.F"),
  c("group2", "S.21D.O.M", "S.21D.Y.M"),
  
  c("group2", "E.2D.Y.F", "E.2D.Y.M"),
  c("group2", "E.2D.O.F", "E.2D.O.M"),
  c("group2", "S.2D.Y.F", "S.2D.Y.M"),
  c("group2", "S.2D.O.F", "S.2D.O.M"),
  c("group2", "E.21D.Y.F", "E.21D.Y.M"),
  c("group2", "E.21D.O.F", "E.21D.O.M"),
  c("group2", "S.21D.Y.F", "S.21D.Y.M"),
  c("group2", "S.21D.O.F", "S.21D.O.M"),
  
  # E - S comparisons
  c("group", "E.2D.Y", "S.2D.Y"),
  c("group", "E.21D.Y", "S.21D.Y"),
  c("group", "E.2D.O", "S.2D.O"),
  c("group", "E.21D.O", "S.21D.O"),

  # 21D - 2D comparisons
  c("group", "E.21D.Y", "E.2D.Y"),
  c("group", "S.21D.Y", "S.2D.Y"),
  c("group", "E.21D.O", "E.2D.O"),
  c("group", "S.21D.O", "S.2D.O"),

  # O - Y comparisons
  c("group", "E.2D.O", "E.2D.Y"),
  c("group", "S.2D.O", "S.2D.Y"),
  c("group", "E.21D.O", "E.21D.Y"),
  c("group", "S.21D.O", "S.21D.Y")
)

```

## DE for all cell types
```{r de_all}
for (cmp in comparisons) {
  group_col <- cmp[1]
  g1 <- cmp[2]
  g2 <- cmp[3]

  # Filter metadata for current comparison
  meta_cmp <- sample_meta %>%
    filter(.data[[group_col]] %in% c(g1, g2)) %>%
    distinct(sample, .keep_all = TRUE)

  comparison_results <- list()

  for (ct in cell_types) {
    
    # get cell type counts matrix
    sample_cols <- paste0(ct, "_", meta_cmp$sample)
    pb_counts <- split_pb_counts[[ct]]
    
    # subset cell type matrix based on samples of interest for comparison
    pb_counts <- pb_counts[, colnames(split_pb_counts[[ct]]) %in% sample_cols, drop = FALSE]

    # DESeq2
    res <- DESeq2_per_cell_type(
      cell_type = ct,
      pb_counts = pb_counts,
      sample_meta = meta_cmp,
      group_column = group_col,
      group1 = g1,
      group2 = g2,
      min_counts = 20,
      samples_per_group = NULL
    )

    if (!is.null(res)) {
      comparison_results[[ct]] <- res
    }
  }

  # Combine all cell types for this comparison and write once
  all_res <- do.call(rbind, comparison_results)
  output_file <- file.path(paste0(out, "DEGs_pseudobulk/DEG_tables/", 
                                  g1, "_vs_", g2, "_DEGs.tsv"))
  write.table(all_res,
              file = output_file, 
              sep = "\t", 
              quote = FALSE, 
              row.names = FALSE)
}
```


# DEG Plots
## Sex Specific Comparisons
### Female vs Male
```{r compare_degs}
# set thresh
adjpval_thresh <- 0.05
lfc_thresh <- 0

# get file paths
files <- list.files(paste0(out, "DEGs_pseudobulk/DEG_tables"))
files <- files[grepl(".+F_vs_.+M_DEGs.tsv", files)]

# get comparisons
comparisons <- gsub("_DEGs.tsv", "", files)
all_comparisons <- c(paste0(comparisons, "_up"),
                     paste0(comparisons, "_down"))

# get cell types
cell_types <- unique(mouse_unannotated$seurat_clusters)

# init deg.df
deg.df <- data.frame(
  matrix(0, nrow = length(all_comparisons), ncol = length(cell_types)),
  row.names = all_comparisons
)
colnames(deg.df) <- cell_types

# reorder rows
deg.df <- deg.df[c(16,8,12,4,15,7,11,3,14,6,10,2,13,5,9,1),]

# loop through files
for (i in files) {
  
  # read table
  df <- read.table(paste0(out, "DEGs_pseudobulk/DEG_tables/", i),
                   sep = "\t", 
                   header = TRUE)
  
  # filter
  df <- df[!is.na(df$padj),]
  df <- df[df$padj < adjpval_thresh,]
  df <- df[df$log2FoldChange < -lfc_thresh | df$log2FoldChange > lfc_thresh,]
  
  # add direction column
  comparisonName <- gsub("_DEGs.tsv", "", i)
  direction <- df$log2FoldChange > 0 # already filtered on lfc so only matters it's positive/negative
  direction <- gsub(TRUE, paste0(comparisonName, "_up"), direction)
  direction <- gsub(FALSE, paste0(comparisonName, "_down"), direction)
  df$direction <- direction
  
  # reformat table
  df <- df %>% 
    dplyr::count(cell_type, direction) %>%
    tidyr::spread(cell_type, n)
  rownames(df) <- df$direction
  df$direction <- NULL
  
  # add to master table
  deg.df[rownames(df), colnames(df)] <- df
}

# reformat
deg.df[is.na(deg.df)] <- 0

# save
prefix <- paste0(out, "DEGs_pseudobulk/DEG_comparison/",
                 "female_vs_male_DEG_sex_specific_comparison_FDRq_", 
                 format(adjpval_thresh, nsmall = 2),
                 "_LFC_", format(lfc_thresh, nsmall = 2))
write.table(df, 
            paste0(prefix, ".tsv"),
            sep = "\t", quote = FALSE, row.names = TRUE)

# set heatmap colors and names
paletteLength <- 100
myColor <- colorRampPalette(c("white","#f0eb9e","darkgreen"))(paletteLength)

meta <- data.frame(treatment = rep(c("Saline", "Saline", "E. coli", "E. coli"), 4),
                   age = c(rep("Young", 4), rep("Old", 4),
                           rep("Young", 4), rep("Old", 4)),
                   time_point = c(rep("2 days", 8), rep("21 days", 8)),
                   direction = rep(c("Down-regulated","Up-regulated"), 8))
rownames(meta) <- rownames(deg.df)

ann_colors <- list(treatment = c(`E. coli` = "pink2", Saline = "chocolate4"),
                   age = c(Young = "gold",
                           Old = "chartreuse2"),
                   time_point = c(`2 days` = "cyan", 
                                  `21 days` = "purple"),
                   direction = c(`Down-regulated` = "#0271f0",
                                 `Up-regulated` = "tomato"))

# save
pdf(paste0(prefix, ".pdf"), width = 10, height = 10)

# plot
pheatmap::pheatmap(deg.df,
                   main = paste0("Female vs Male, FDRq < ", format(adjpval_thresh, nsmall = 2), 
                                 ", |LFC| > ", format(lfc_thresh, nsmall = 2)),
                   treeheight_row = 0,
                   treeheight_col = 0,
                   color = myColor,
                   cluster_rows = FALSE,
                   annotation_row = meta,
                   annotation_colors = ann_colors,
                   display_numbers = round(deg.df, digits = 0),
                   fontsize_number = 12,
                   number_color = "black")
```

### Old vs Young
```{r old_vs_young_degs, eval=FALSE}
# set thresh
adjpval_thresh <- 0.05
lfc_thresh <- 0

# get file paths
files <- list.files(paste0(out, "DEGs_pseudobulk/DEG_tables"))
files <- files[grepl(".+O.+_vs_.+Y.+_DEGs.tsv", files)]

# get comparisons
comparisons <- gsub("_DEGs.tsv", "", files)
all_comparisons <- c(paste0(comparisons, "_up"),
                     paste0(comparisons, "_down"))

# get cell types
cell_types <- unique(mouse_unannotated$seurat_clusters)

# init deg.df
deg.df <- data.frame(
  matrix(0, nrow = length(all_comparisons), ncol = length(cell_types)),
  row.names = all_comparisons
)
colnames(deg.df) <- cell_types

# reorder rows
deg.df <- deg.df[c(15,7,11,3,16,8,12,4,13,5,9,1,14,6,10,2),]

# loop through files
for (i in files) {
  
  # read table
  df <- read.table(paste0(out, "DEGs_pseudobulk/DEG_tables/", i),
                   sep = "\t", 
                   header = TRUE)
  
  # filter
  df <- df[!is.na(df$padj),]
  df <- df[df$padj < adjpval_thresh,]
  df <- df[df$log2FoldChange < -lfc_thresh | df$log2FoldChange > lfc_thresh,]
  
  # add direction column
  comparisonName <- gsub("_DEGs.tsv", "", i)
  direction <- df$log2FoldChange > 0 # already filtered on lfc so only matters it's positive/negative
  direction <- gsub(TRUE, paste0(comparisonName, "_up"), direction)
  direction <- gsub(FALSE, paste0(comparisonName, "_down"), direction)
  df$direction <- direction
  
  # reformat table
  df <- df %>% 
    dplyr::count(cell_type, direction) %>%
    tidyr::spread(cell_type, n)
  rownames(df) <- df$direction
  df$direction <- NULL
  
  # add to master table
  deg.df[rownames(df), colnames(df)] <- df
}

# reformat
deg.df[is.na(deg.df)] <- 0

# save
prefix <- paste0(out, "DEGs_pseudobulk/DEG_comparison/",
                 "old_vs_young_DEG_sex_specific_comparison_FDRq_", 
                 format(adjpval_thresh, nsmall = 2),
                 "_LFC_", format(lfc_thresh, nsmall = 2))
write.table(df, 
            paste0(prefix, ".tsv"),
            sep = "\t", quote = FALSE, row.names = TRUE)

# set heatmap colors and names
paletteLength <- 100
myColor <- colorRampPalette(c("white","#f0eb9e","darkgreen"))(paletteLength)

meta <- data.frame(treatment = rep(c("Saline", "Saline", "E. coli", "E. coli"), 4),
                   sex = c(rep("Female", 4), rep("Male", 4),
                           rep("Female", 4), rep("Male", 4)),
                   time_point = c(rep("2 days", 8), rep("21 days", 8)),
                   direction = rep(c("Down-regulated","Up-regulated"), 8))
rownames(meta) <- rownames(deg.df)

ann_colors <- list(treatment = c(`E. coli` = "pink2", Saline = "chocolate4"),
                   sex = c(Female = "gray",
                           Male = "orange"),
                   time_point = c(`2 days` = "cyan", 
                                  `21 days` = "purple"),
                   direction = c(`Down-regulated` = "#0271f0",
                                 `Up-regulated` = "tomato"))

# save
pdf(paste0(prefix, ".pdf"), width = 10, height = 10)

# plot
pheatmap::pheatmap(deg.df,
                   main = paste0("Old vs Young, FDRq < ", format(adjpval_thresh, nsmall = 2), 
                                 ", |LFC| > ", format(lfc_thresh, nsmall = 2)),
                   treeheight_row = 0,
                   treeheight_col = 0,
                   color = myColor,
                   cluster_rows = FALSE,
                   annotation_row = meta,
                   annotation_colors = ann_colors,
                   display_numbers = round(deg.df, digits = 0),
                   fontsize_number = 12,
                   number_color = "black")
```

### 21 vs 2 days
```{r old_vs_young_degs, eval=FALSE}
# set thresh
adjpval_thresh <- 0.05
lfc_thresh <- 0

# get file paths
files <- list.files(paste0(out, "DEGs_pseudobulk/DEG_tables"))
files <- files[grepl(".+21D.+[MF]_vs_.+2D.+[MF]_DEGs.tsv", files)]

# get comparisons
comparisons <- gsub("_DEGs.tsv", "", files)
all_comparisons <- c(paste0(comparisons, "_up"),
                     paste0(comparisons, "_down"))

# get cell types
cell_types <- unique(mouse_unannotated$seurat_clusters)

# init deg.df
deg.df <- data.frame(
  matrix(0, nrow = length(all_comparisons), ncol = length(cell_types)),
  row.names = all_comparisons
)
colnames(deg.df) <- cell_types

# reorder rows
deg.df <- deg.df[c(15,7,11,3,16,8,12,4,13,5,9,1,14,6,10,2),]

# loop through files
for (i in files) {
  
  # read table
  df <- read.table(paste0(out, "DEGs_pseudobulk/DEG_tables/", i),
                   sep = "\t", 
                   header = TRUE)
  
  # filter
  df <- df[!is.na(df$padj),]
  df <- df[df$padj < adjpval_thresh,]
  df <- df[df$log2FoldChange < -lfc_thresh | df$log2FoldChange > lfc_thresh,]
  
  # add direction column
  comparisonName <- gsub("_DEGs.tsv", "", i)
  direction <- df$log2FoldChange > 0 # already filtered on lfc so only matters it's positive/negative
  direction <- gsub(TRUE, paste0(comparisonName, "_up"), direction)
  direction <- gsub(FALSE, paste0(comparisonName, "_down"), direction)
  df$direction <- direction
  
  # reformat table
  df <- df %>% 
    dplyr::count(cell_type, direction) %>%
    tidyr::spread(cell_type, n)
  rownames(df) <- df$direction
  df$direction <- NULL
  
  # add to master table
  deg.df[rownames(df), colnames(df)] <- df
}

# reformat
deg.df[is.na(deg.df)] <- 0

# save
prefix <- paste0(out, "DEGs_pseudobulk/DEG_comparison/",
                 "21_vs_2_days_DEG_sex_specific_comparison_FDRq_", 
                 format(adjpval_thresh, nsmall = 2),
                 "_LFC_", format(lfc_thresh, nsmall = 2))
write.table(df, 
            paste0(prefix, ".tsv"),
            sep = "\t", quote = FALSE, row.names = TRUE)

# set heatmap colors and names
paletteLength <- 100
myColor <- colorRampPalette(c("white","#f0eb9e","darkgreen"))(paletteLength)

meta <- data.frame(treatment = rep(c("Saline", "Saline", "E. coli", "E. coli"), 4),
                   sex = c(rep("Female", 4), rep("Male", 4),
                           rep("Female", 4), rep("Male", 4)),
                   age = c(rep("Young", 8), rep("Old", 8)),
                   direction = rep(c("Down-regulated","Up-regulated"), 8))
rownames(meta) <- rownames(deg.df)

ann_colors <- list(treatment = c(`E. coli` = "pink2", Saline = "chocolate4"),
                   sex = c(Female = "gray",
                           Male = "orange"),
                   age = c(Young = "gold",
                           Old = "chartreuse2"),
                   direction = c(`Down-regulated` = "#0271f0",
                                 `Up-regulated` = "tomato"))

# save
pdf(paste0(prefix, ".pdf"), width = 10, height = 10)

# plot
pheatmap::pheatmap(deg.df,
                   main = paste0("21 vs 2 Days, FDRq < ", format(adjpval_thresh, nsmall = 2), 
                                 ", |LFC| > ", format(lfc_thresh, nsmall = 2)),
                   treeheight_row = 0,
                   treeheight_col = 0,
                   color = myColor,
                   cluster_rows = FALSE,
                   annotation_row = meta,
                   annotation_colors = ann_colors,
                   display_numbers = round(deg.df, digits = 0),
                   fontsize_number = 12,
                   number_color = "black")
```


### Ecoli vs saline
```{r old_vs_young_degs, eval=FALSE}
# set thresh
adjpval_thresh <- 0.05
lfc_thresh <- 0

# get file paths
files <- list.files(paste0(out, "DEGs_pseudobulk/DEG_tables"))
files <- files[grepl("E.+[MF]_vs_S.+[MF]_DEGs.tsv", files)]

# get comparisons
comparisons <- gsub("_DEGs.tsv", "", files)
all_comparisons <- c(paste0(comparisons, "_up"),
                     paste0(comparisons, "_down"))

# get cell types
cell_types <- unique(mouse_unannotated$seurat_clusters)

# init deg.df
deg.df <- data.frame(
  matrix(0, nrow = length(all_comparisons), ncol = length(cell_types)),
  row.names = all_comparisons
)
colnames(deg.df) <- cell_types

# reorder rows
deg.df <- deg.df[c(15,7,16,8,13,5,14,6,11,3,12,4,9,1,10,2),]

# loop through files
for (i in files) {
  
  # read table
  df <- read.table(paste0(out, "DEGs_pseudobulk/DEG_tables/", i),
                   sep = "\t", 
                   header = TRUE)
  
  # filter
  df <- df[!is.na(df$padj),]
  df <- df[df$padj < adjpval_thresh,]
  df <- df[df$log2FoldChange < -lfc_thresh | df$log2FoldChange > lfc_thresh,]
  
  # add direction column
  comparisonName <- gsub("_DEGs.tsv", "", i)
  direction <- df$log2FoldChange > 0 # already filtered on lfc so only matters it's positive/negative
  direction <- gsub(TRUE, paste0(comparisonName, "_up"), direction)
  direction <- gsub(FALSE, paste0(comparisonName, "_down"), direction)
  df$direction <- direction
  
  # reformat table
  df <- df %>% 
    dplyr::count(cell_type, direction) %>%
    tidyr::spread(cell_type, n)
  rownames(df) <- df$direction
  df$direction <- NULL
  
  # add to master table
  deg.df[rownames(df), colnames(df)] <- df
}

# reformat
deg.df[is.na(deg.df)] <- 0

# save
prefix <- paste0(out, "DEGs_pseudobulk/DEG_comparison/",
                 "ecoli_vs_saline_DEG_sex_specific_comparison_FDRq_", 
                 format(adjpval_thresh, nsmall = 2),
                 "_LFC_", format(lfc_thresh, nsmall = 2))
write.table(df, 
            paste0(prefix, ".tsv"),
            sep = "\t", quote = FALSE, row.names = TRUE)

# set heatmap colors and names
paletteLength <- 100
myColor <- colorRampPalette(c("white","#f0eb9e","darkgreen"))(paletteLength)

meta <- data.frame(time_point =  c(rep("2 days", 8), rep("21 days", 8)),
                   sex = rep(c("Female","Female","Male","Male"), 4),
                   age = c(rep("Young", 4), rep("Old", 4),
                           rep("Young", 4), rep("Old", 4)),
                   direction = rep(c("Down-regulated","Up-regulated"), 8))
rownames(meta) <- rownames(deg.df)

ann_colors <- list(time_point = c(`2 days` = "cyan", 
                                  `21 days` = "purple"),
                   sex = c(Female = "gray",
                           Male = "orange"),
                   age = c(Young = "gold",
                           Old = "chartreuse2"),
                   direction = c(`Down-regulated` = "#0271f0",
                                 `Up-regulated` = "tomato"))

# save
pdf(paste0(prefix, ".pdf"), width = 10, height = 10)

# plot
pheatmap::pheatmap(deg.df,
                   main = paste0("E. coli vs Saline, FDRq < ", format(adjpval_thresh, nsmall = 2), 
                                 ", |LFC| > ", format(lfc_thresh, nsmall = 2)),
                   treeheight_row = 0,
                   treeheight_col = 0,
                   color = myColor,
                   cluster_rows = FALSE,
                   annotation_row = meta,
                   annotation_colors = ann_colors,
                   display_numbers = round(deg.df, digits = 0),
                   fontsize_number = 12,
                   number_color = "black")
```

## Both Sexes Comparisons
### Old vs Young
```{r old_vs_young_degs, eval=FALSE}
# set thresh
adjpval_thresh <- 0.05
lfc_thresh <- 0

# get file paths
files <- list.files(paste0(out, "DEGs_pseudobulk/DEG_tables"))
files <- files[grepl(".+O_vs_.+Y_DEGs.tsv", files)]

# get comparisons
comparisons <- gsub("_DEGs.tsv", "", files)
all_comparisons <- c(paste0(comparisons, "_up"),
                     paste0(comparisons, "_down"))

# get cell types
cell_types <- unique(mouse_unannotated$seurat_clusters)

# init deg.df
deg.df <- data.frame(
  matrix(0, nrow = length(all_comparisons), ncol = length(cell_types)),
  row.names = all_comparisons
)
colnames(deg.df) <- cell_types

# reorder rows
deg.df <- deg.df[c(8,4,6,2,7,3,5,1),]

# loop through files
for (i in files) {
  
  # read table
  df <- read.table(paste0(out, "DEGs_pseudobulk/DEG_tables/", i),
                   sep = "\t", 
                   header = TRUE)
  
  # filter
  df <- df[!is.na(df$padj),]
  df <- df[df$padj < adjpval_thresh,]
  df <- df[df$log2FoldChange < -lfc_thresh | df$log2FoldChange > lfc_thresh,]
  
  # add direction column
  comparisonName <- gsub("_DEGs.tsv", "", i)
  direction <- df$log2FoldChange > 0 # already filtered on lfc so only matters it's positive/negative
  direction <- gsub(TRUE, paste0(comparisonName, "_up"), direction)
  direction <- gsub(FALSE, paste0(comparisonName, "_down"), direction)
  df$direction <- direction
  
  # reformat table
  df <- df %>% 
    dplyr::count(cell_type, direction) %>%
    tidyr::spread(cell_type, n)
  rownames(df) <- df$direction
  df$direction <- NULL
  
  # add to master table
  deg.df[rownames(df), colnames(df)] <- df
}

# reformat
deg.df[is.na(deg.df)] <- 0

# save
prefix <- paste0(out, "DEGs_pseudobulk/DEG_comparison/",
                 "old_vs_young_DEG_both_sexes_comparison_FDRq_", 
                 format(adjpval_thresh, nsmall = 2),
                 "_LFC_", format(lfc_thresh, nsmall = 2))
write.table(df, 
            paste0(prefix, ".tsv"),
            sep = "\t", quote = FALSE, row.names = TRUE)

# set heatmap colors and names
paletteLength <- 100
myColor <- colorRampPalette(c("white","#f0eb9e","darkgreen"))(paletteLength)

meta <- data.frame(treatment = rep(c("Saline", "Saline", "E. coli", "E. coli"), 2),
                   time_point = c(rep("2 days", 4), rep("21 days", 4)),
                   direction = rep(c("Down-regulated","Up-regulated"), 4))
rownames(meta) <- rownames(deg.df)

ann_colors <- list(treatment = c(`E. coli` = "pink2", Saline = "chocolate4"),
                   time_point = c(`2 days` = "cyan", 
                                  `21 days` = "purple"),
                   direction = c(`Down-regulated` = "#0271f0",
                                 `Up-regulated` = "tomato"))

# save
pdf(paste0(prefix, ".pdf"), width = 10, height = 6)

# plot
pheatmap::pheatmap(deg.df,
                   main = paste0("Old vs Young, FDRq < ", format(adjpval_thresh, nsmall = 2), 
                                 ", |LFC| > ", format(lfc_thresh, nsmall = 2)),
                   treeheight_row = 0,
                   treeheight_col = 0,
                   color = myColor,
                   cluster_rows = FALSE,
                   annotation_row = meta,
                   annotation_colors = ann_colors,
                   display_numbers = round(deg.df, digits = 0),
                   fontsize_number = 12,
                   number_color = "black")
```

### 21 vs 2 days
```{r old_vs_young_degs, eval=FALSE}
# set thresh
adjpval_thresh <- 0.05
lfc_thresh <- 0

# get file paths
files <- list.files(paste0(out, "DEGs_pseudobulk/DEG_tables"))
files <- files[grepl(".+21D.+[OY]_vs_.+2D.+[OY]_DEGs.tsv", files)]

# get comparisons
comparisons <- gsub("_DEGs.tsv", "", files)
all_comparisons <- c(paste0(comparisons, "_up"),
                     paste0(comparisons, "_down"))

# get cell types
cell_types <- unique(mouse_unannotated$seurat_clusters)

# init deg.df
deg.df <- data.frame(
  matrix(0, nrow = length(all_comparisons), ncol = length(cell_types)),
  row.names = all_comparisons
)
colnames(deg.df) <- cell_types

# reorder rows
deg.df <- deg.df[c(8,4,6,2,7,3,5,1),]

# loop through files
for (i in files) {
  
  # read table
  df <- read.table(paste0(out, "DEGs_pseudobulk/DEG_tables/", i),
                   sep = "\t", 
                   header = TRUE)
  
  # filter
  df <- df[!is.na(df$padj),]
  df <- df[df$padj < adjpval_thresh,]
  df <- df[df$log2FoldChange < -lfc_thresh | df$log2FoldChange > lfc_thresh,]
  
  # add direction column
  comparisonName <- gsub("_DEGs.tsv", "", i)
  direction <- df$log2FoldChange > 0 # already filtered on lfc so only matters it's positive/negative
  direction <- gsub(TRUE, paste0(comparisonName, "_up"), direction)
  direction <- gsub(FALSE, paste0(comparisonName, "_down"), direction)
  df$direction <- direction
  
  # reformat table
  df <- df %>% 
    dplyr::count(cell_type, direction) %>%
    tidyr::spread(cell_type, n)
  rownames(df) <- df$direction
  df$direction <- NULL
  
  # add to master table
  deg.df[rownames(df), colnames(df)] <- df
}

# reformat
deg.df[is.na(deg.df)] <- 0

# save
prefix <- paste0(out, "DEGs_pseudobulk/DEG_comparison/",
                 "21_vs_2_days_DEG_both_sexes_comparison_FDRq_", 
                 format(adjpval_thresh, nsmall = 2),
                 "_LFC_", format(lfc_thresh, nsmall = 2))
write.table(df, 
            paste0(prefix, ".tsv"),
            sep = "\t", quote = FALSE, row.names = TRUE)

# set heatmap colors and names
paletteLength <- 100
myColor <- colorRampPalette(c("white","#f0eb9e","darkgreen"))(paletteLength)

meta <- data.frame(treatment = rep(c("Saline", "Saline", "E. coli", "E. coli"), 2),
                   age = c(rep("Young", 4), rep("Old", 4)),
                   direction = rep(c("Down-regulated","Up-regulated"), 4))
rownames(meta) <- rownames(deg.df)

ann_colors <- list(treatment = c(`E. coli` = "pink2", Saline = "chocolate4"),
                   sex = c(Female = "gray",
                           Male = "orange"),
                   age = c(Young = "gold",
                           Old = "chartreuse2"),
                   direction = c(`Down-regulated` = "#0271f0",
                                 `Up-regulated` = "tomato"))

# save
pdf(paste0(prefix, ".pdf"), width = 10, height = 6)

# plot
pheatmap::pheatmap(deg.df,
                   main = paste0("21 vs 2 Days, FDRq < ", format(adjpval_thresh, nsmall = 2), 
                                 ", |LFC| > ", format(lfc_thresh, nsmall = 2)),
                   treeheight_row = 0,
                   treeheight_col = 0,
                   color = myColor,
                   cluster_rows = FALSE,
                   annotation_row = meta,
                   annotation_colors = ann_colors,
                   display_numbers = round(deg.df, digits = 0),
                   fontsize_number = 12,
                   number_color = "black")
```


### Ecoli vs saline
```{r both_sexes_ecoli_vs_saline_degs, eval=FALSE}
# set thresh
adjpval_thresh <- 0.05
lfc_thresh <- 0

# get file paths
files <- list.files(paste0(out, "DEGs_pseudobulk/DEG_tables"))
files <- files[grepl("E.+[OY]_vs_S.+[OY]_DEGs.tsv", files)]

# get comparisons
comparisons <- gsub("_DEGs.tsv", "", files)
all_comparisons <- c(paste0(comparisons, "_up"),
                     paste0(comparisons, "_down"))

# get cell types
cell_types <- unique(mouse_unannotated$seurat_clusters)

# init deg.df
deg.df <- data.frame(
  matrix(0, nrow = length(all_comparisons), ncol = length(cell_types)),
  row.names = all_comparisons
)
colnames(deg.df) <- cell_types

# reorder rows
deg.df <- deg.df[c(8,4,7,3,6,2,5,1),]

# loop through files
for (i in files) {
  
  # read table
  df <- read.table(paste0(out, "DEGs_pseudobulk/DEG_tables/", i),
                   sep = "\t", 
                   header = TRUE)
  
  # filter
  df <- df[!is.na(df$padj),]
  df <- df[df$padj < adjpval_thresh,]
  df <- df[df$log2FoldChange < -lfc_thresh | df$log2FoldChange > lfc_thresh,]
  
  # add direction column
  comparisonName <- gsub("_DEGs.tsv", "", i)
  direction <- df$log2FoldChange > 0 # already filtered on lfc so only matters it's positive/negative
  direction <- gsub(TRUE, paste0(comparisonName, "_up"), direction)
  direction <- gsub(FALSE, paste0(comparisonName, "_down"), direction)
  df$direction <- direction
  
  # reformat table
  df <- df %>% 
    dplyr::count(cell_type, direction) %>%
    tidyr::spread(cell_type, n)
  rownames(df) <- df$direction
  df$direction <- NULL
  
  # add to master table
  deg.df[rownames(df), colnames(df)] <- df
}

# reformat
deg.df[is.na(deg.df)] <- 0

# save
prefix <- paste0(out, "DEGs_pseudobulk/DEG_comparison/",
                 "ecoli_vs_saline_DEG_both_sexes_comparison_FDRq_", 
                 format(adjpval_thresh, nsmall = 2),
                 "_LFC_", format(lfc_thresh, nsmall = 2))
write.table(df, 
            paste0(prefix, ".tsv"),
            sep = "\t", quote = FALSE, row.names = TRUE)

# set heatmap colors and names
paletteLength <- 100
myColor <- colorRampPalette(c("white","#f0eb9e","darkgreen"))(paletteLength)

meta <- data.frame(time_point =  c(rep("2 days", 4), rep("21 days", 4)),
                   age = c(rep("Young", 2), rep("Old", 2),
                           rep("Young", 2), rep("Old", 2)),
                   direction = rep(c("Down-regulated","Up-regulated"), 4))
rownames(meta) <- rownames(deg.df)

ann_colors <- list(time_point = c(`2 days` = "cyan", 
                                  `21 days` = "purple"),
                   age = c(Young = "gold",
                           Old = "chartreuse2"),
                   direction = c(`Down-regulated` = "#0271f0",
                                 `Up-regulated` = "tomato"))

# save
pdf(paste0(prefix, ".pdf"), width = 10, height = 6)

# plot
pheatmap::pheatmap(deg.df,
                   main = paste0("E. coli vs Saline, FDRq < ", format(adjpval_thresh, nsmall = 2), 
                                 ", |LFC| > ", format(lfc_thresh, nsmall = 2)),
                   treeheight_row = 0,
                   treeheight_col = 0,
                   color = myColor,
                   cluster_rows = FALSE,
                   annotation_row = meta,
                   annotation_colors = ann_colors,
                   display_numbers = round(deg.df, digits = 0),
                   fontsize_number = 12,
                   number_color = "black")
```

## Volcano
```{r volcano, message=FALSE, eval=FALSE}
# set thresh
adjpval_thresh <- 0.05
lfc_thresh <- 0

# get file paths
files <- list.files(paste0(out, "DEGs_pseudobulk/DEG_tables"), full.names = TRUE)
files <- files[grepl(".+_DEGs.tsv", files)]

for (file in files) {
  
  # read DEG file
  treatment_vs_control <- read.delim(file)
  
  # filter NA values
  treatment_vs_control <- treatment_vs_control[!is.na(treatment_vs_control$padj),]
  
  # extract title for plot later
  p.comparison <- str_match(file, "DEG_tables/(.+)_DEGs.tsv")[,2]
  
  # assign colors
  color_values <- vector()
  max <- nrow(treatment_vs_control)
  for(row in 1:max){
    if (treatment_vs_control$padj[row] < adjpval_thresh){
      if (treatment_vs_control$log2FoldChange [row] > lfc_thresh){
        color_values <- c(color_values, 1) # 1 when logFC > thresh and FDRq < thresh
      } else if (treatment_vs_control$log2FoldChange[row] < -lfc_thresh){
        color_values <- c(color_values, 2) # 2 when logFC < thresh and FDRq < thresh
      } else {
        color_values <- c(color_values, 3) # 2 when FDRq < thresh and LFC not met
      }
    }
    else{
      color_values <- c(color_values, 3) # 3 when FDRq >= thresh
    }
  }
  treatment_vs_control$color_adjpval_thresh <- factor(color_values)
  
  # loop through clusters
  for (j in unique(treatment_vs_control$cell_type)) {
    
    # subset cluster
    data <- subset(treatment_vs_control, cell_type == j)
    
    # plot only if there are DEGs with padj < thresh
    num <- subset(data, padj < adjpval_thresh)
    num <- nrow(num)
    if(num != 0) {
        
      # subset genes to label
      up <- data[data$color_adjpval_thresh == 1,]
      up.sig <- up[order(up$padj),][1:15,]
      up.lfc <- up[order(up$log2FoldChange, decreasing = TRUE),][1:15,]
      up30 <- rbind(up.sig,up.lfc)
      up30 <- unique(up30)
      up30 <- up30[rowSums(is.na(up30)) != ncol(up30), ]
      down <- data[data$color_adjpval_thresh == 2,]
      down.sig <- down[order(down$padj),][1:15,]
      down.lfc <- down[order(down$log2FoldChange, decreasing = FALSE),][1:15,]
      down30 <- rbind(down.sig,down.lfc)
      down30 <- unique(down30)
      down30 <- down30[rowSums(is.na(down30)) != ncol(down30), ]
      
      # set manual colors
      if (!1 %in% unique(data$color_adjpval_thresh)) {
        my_colors <- c("blue","gray")
      } else if (!2 %in% unique(data$color_adjpval_thresh)) {
        my_colors <- c("red","gray")
      } else if (!1 %in% unique(data$color_adjpval_thresh) && !2 %in% unique(data$color_adjpval_thresh)) {
        my_colors <- c("gray")
      } else {
        my_colors <- c("red","blue","gray")
      }
      
      # set significance threshold
      hadjpval <- (-log10(max(
        data$pvalue[data$padj < adjpval_thresh], 
        na.rm=TRUE)))

      # plot
      p <-
        ggplot(data = data, 
               aes(x = log2FoldChange,     # x-axis is logFC
                   y = -log10(pvalue),  # y-axis will be -log10 of P.Value
                   color = color_adjpval_thresh)) +  # color is based on factored color column
        geom_point(alpha = 0.8, size = 2) +  # create scatterplot, alpha makes points transparent
        theme_bw() +  # set color theme
        theme(legend.position = "none") +  # no legend
        scale_color_manual(values = my_colors) +
        labs(
          title = "",
          x = expression(log[2](FC)),
          y = expression(-log[10] ~ "(" ~ italic("p") ~ "-value)")
        ) +
        theme(axis.title.x = element_text(size = 10),
              axis.text.x = element_text(size = 10)) +
        theme(axis.title.y = element_text(size = 10),
              axis.text.y = element_text(size = 10)) +
        geom_hline(yintercept = hadjpval, colour = "#000000", linetype = "dashed") +
        geom_vline(xintercept = lfc_thresh, colour = "#000000", linetype = "dashed") +
        geom_vline(xintercept = -lfc_thresh, colour = "#000000", linetype = "dashed") +
        ggtitle(paste0(j, "\n", p.comparison, ", q < ", format(adjpval_thresh, nsmall = 2), 
                       ", |LFC| > ", format(lfc_thresh, nsmall = 2))) +
        geom_text_repel(data = up30,
                        aes(x = log2FoldChange, y= -log10(pvalue), label = gene), 
                        color = "maroon", 
                        fontface="italic",
                        max.overlaps = getOption("ggrepel.max.overlaps", default = 30)
                        ) +
        geom_text_repel(data = down30,
                        aes(x = log2FoldChange, y= -log10(pvalue), label = gene), 
                        color = "navyblue", 
                        fontface="italic",
                        max.overlaps = getOption("ggrepel.max.overlaps", default = 30)
                        )
      p
      
      # create dir if doesn't exist
      prefix <- paste0(out, "DEGs_pseudobulk/volcano/", p.comparison)
      if (!dir.exists(prefix)) {
        dir.create(prefix, recursive = TRUE)
      }
      
      # save
      j_reformat <- gsub(" / ", "_", j)
      j_reformat <- gsub(" ", "_", j_reformat)
      plot_out <- paste0(prefix, "/", p.comparison, "_", tolower(j_reformat),
                    "_FDRq_", format(adjpval_thresh, nsmall = 2), "_LFC_", 
                    format(lfc_thresh, nsmall = 2), "_volcano.pdf")
      pdf(plot_out, height = 8, width = 8)
      print(p)
      dev.off()
      
      print(paste("comparison =",p.comparison,", cell type =", j))
      
    }
  } # end loop through clusters
} # end loop through variables
```

## Metascape input
```{r metascape_sex_specific}
# set thresh
adjpval_thresh <- 0.05
lfc_thresh <- 0

# get file paths
files <- list.files(paste0(out, "DEGs_pseudobulk/DEG_tables"), full.names = TRUE)
files <- files[grepl(".+_DEGs.tsv", files)]

# loop through files
for (i in 1:length(files)) {
  
  # read table
  data <- read.table(files[i], header = TRUE, sep = "\t")
  
  # filter
  data <- data[!is.na(data$padj),]
  data <- data[data$padj < adjpval_thresh,]
  data <- data[data$log2FoldChange < -lfc_thresh | data$log2FoldChange > lfc_thresh,]
  
  # separate into up and down-regulated
  up.df <- data[data$log2FoldChange > 0,]
  down.df <- data[data$log2FoldChange < 0,]
  
  cell_types <- unique(data$cell_type)
  
  # create dir if doesn't exist
  p.comparison <- str_match(files[i], "DEG_tables/(.+)_DEGs.tsv")[,2]
  prefix <- paste0(out, "DEGs_pseudobulk/metascape_input/", p.comparison)
  if (!dir.exists(prefix)) {
    dir.create(prefix, recursive = TRUE)
  }
  
  # loop through cell_type
  for (j in cell_types) {
    
    # subset based on cell_type
    print(paste0(p.comparison, ": ", j))
    up <- up.df[up.df$cell_type == j,]
    up <- up$gene
    down <- down.df[down.df$cell_type == j,]
    down <- down$gene
    
    # file name
    prefix <- paste0(out, "DEGs_pseudobulk/metascape_input/", p.comparison,
                     "/", p.comparison, "_", gsub(" ", "_", tolower(j)), 
                     "_FDRq_", format(adjpval_thresh, nsmall = 2), 
                     "_LFC_", format(lfc_thresh, nsmall = 2))
     
    # save
    if(length(up) > 10) {
      write.table(x = up,
                  file = paste0(prefix, "_up.txt"),
                  quote = FALSE,
                  row.names = FALSE,
                  col.names = FALSE)      
    }
    if (length(down) > 10) {
       write.table(x = down,
                file = paste0(prefix, "_down.txt"),
                quote = FALSE,
                row.names = FALSE,
                col.names = FALSE)   
    }
  } # end j for loop
} # end i for loop
```

# Shiny app
```{r shiny_app, eval=FALSE}
# create new object
library(ShinyCell)
shiny.obj <- mouse_unannotated
VariableFeatures(shiny.obj) <- shiny.obj@assays$SCT@var.features

# set default params
DefaultAssay(shiny.obj) <- "RNA"

# create config
colNames <- colnames(shiny.obj@meta.data)
colNames <- colNames[c(29,30,2:24)]
sc.config <- createConfig(obj = shiny.obj,
                          meta.to.include = colNames)

# change wd
setwd(out)

# output shiny app folder
makeShinyApp(obj = shiny.obj, 
             scConf = sc.config, 
             gene.mapping = TRUE,
             shiny.title = "E. coli Mice scRNAseq Microglia Recluster")

# cleanup
remove(sc.config,sc1conf,shiny.obj)
```


